作为**NOAI Simple Guidelines**的完结篇，本教程将系统讲解神经网络与深度学习的核心概念，并带你快速上手PyTorch框架的基本操作。内容全面覆盖第一轮线下笔试中关于神经网络的所有考点，同时延伸至第二轮所必需的PyTorch编程能力。这恰恰是拉开分数差距的关键所在——如果你希望在中学阶段就能参与青年学者杯这类有深度的学术活动，神经网络和深度学习是你无法回避、也是处理复杂任务最强大的工具。在第一轮笔试中，这部分内容将以单项选择题、简答题和代码填空题的形式出现，题型多样，且常涉及较复杂的计算推导。因此，在学习时，请务必重视数学原理的理解，并坚持边学边敲代码，在调试中感受模型的工作机制。把这一块啃透，你不仅能在笔试中稳拿高分，更能为后续的项目实战和研究探索打下扎实的基础。

根据[NOAI 2026学术大纲](https://doc.weixin.qq.com/doc/w3_AP4AVwYpACgCN1VGarE8bSb0xLzvA?scode=ALsAXQdXAAwuRecVzr)，在第一轮线下笔试中，NOAI学术活动对于神经网络和深度学习的要求为：

* 理解关键概念中涉及的机器学习、深度学习方法所适用的典型问题范围。
* **能在文档的帮助下，尝试解决机器学习、深度学习的简单问题，其中编程实践中涉及第三方库的调用，题干中将给出调用方法。**（正是因为这个要求，我们需要提前学习第二轮所必需的PyTorch编程能力。）

神经网络与深度学习涉及的重要概念与方法有：

* 理解感知机(Perceptron)、神经网络(Neural Network, NN)。
* 理解误差反向传播(Back propagation, BP)算法，能描述训练中正向传播和反向传播的过程。
* 理解梯度下降(Gradient Descent)，能对其进行简要描述。
* 理解局部最优(Local Optima)、全局最优(Global Optima)。
* 理解卷积神经网络(Convolutional Neural Network, CNN)，能简要描述网络结构。
* 理解神经网络中常用层的原理和计算过程：卷积层(Convolutional Layer)、池化层(Pooling Layer)、全连接层(Full Connected Layer)。
* 理解神经网络中决策函数(Decision Function)、 激活函数(Activation Function)、 损失函数(Loss Function)，能够描述各个函数的作用。
* 知道动量优化方法(例如Adam、AdamW)的核心思想，能描述其与普通梯度下降的差异。
* 知道大语言模型（LLM）的核心思想。
* 掌握 PyTorch 基础语法，包括张量（Tensor）的创建、基本运算、数据类型转换，以及简单神经网络的定义与初始化方法。

建议学习者在理解概念的基础上，动手进行计算并运行每一段代码，并尝试修改参数观察效果。对于 PyTorch 相关内容，建议在本地安装 PyTorch 环境进行实践。

最后啰嗦一句本教程以及Part2的所有代码均在[玻尔®️ / Bohrium®️平台](https://www.bohrium.com/)运行通过。所有有意向参赛的同学都应当学习如何使用[Bohrium Notebook](https://bohrium-doc.dp.tech/docs/userguide/Notebook/)来进行编程和实验。




>[CAICP](https://drive.weixin.qq.com/s?k=ALsAXQdXAAwa2ISM2E#/preview?fileId=s.1970325044026811.757660456Y0J_f.774930487s4se) 全称为 Certified AI Competence Program（简称 CAICP），中文全称为人工智能能力认证活动；该认证与 NOAI 并列，均为 IOAI 中国区官方活动，其学术大纲与 NOAI 一脉相承，有内在的连续性；与 NOAI 的高难度、高选拔性及主要针对高中生不同，CAICP 定位为人工智能普及性活动，难度较低、更具系统及引导性，覆盖小学、初中及高中学段。CAICP 考察内容主要覆盖人工智能基础概念、Python 编程技能、计算思维三个维度。CAICP 人工智能能力认证活动将分为入门级（Junior）和进阶级（Senior）两个难度级别。学生可以根据自身当前水平选报任一级别。
每个难度级别均包括“第一轮”、“第二轮”、“冬令营”三轮测评认证，前两轮均为在线参与，在第一轮中表现优秀的学生可获得参加第二轮的晋级资格。每年将从 CAICP 第二轮中表现优异的学生中，选拔 100 名左右参加 CAICP 冬令营，具体活动安排，以当年度发布的通知为准。本教程可用于CAICP备考，但是依然需要根据[CAICP 学术大纲](https://drive.weixin.qq.com/s?k=ALsAXQdXAAwa2ISM2E#/preview?fileId=s.1970325044026811.757660456Y0J_f.774930487s4se)补充相应级别的知识点，这里不再累述。

## 第一章 从感知机到神经网络

### 1.1 感知机（Perceptron）

感知机是人工神经网络中最简单的模型，由美国心理学家 Frank Rosenblatt 于 1957 年提出。它是理解现代深度学习的起点，其核心思想可以用一句话概括：**对输入信号进行加权求和，然后通过一个阈值判断来产生输出**。

#### 1.1.1 感知机的数学模型

一个感知机接收 $n$ 个输入信号 $x_1, x_2, \ldots, x_n$，每个输入对应一个权重 $w_1, w_2, \ldots, w_n$，此外还有一个偏置项 $b$（也称为阈值）。感知机的输出 $y$ 由以下公式决定：

$$
z = w_1 x_1 + w_2 x_2 + \cdots + w_n x_n + b = \sum_{i=1}^{n} w_i x_i + b
$$

$$
y = f(z) = \begin{cases} 1, & \text{if } z \geq 0 \\ 0, & \text{if } z < 0 \end{cases}
$$

其中，$f(z)$ 被称为**阶跃函数（Step Function）**，它将加权求和的结果映射为 0 或 1 两个类别。这就是最原始的“决策函数”。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/264c68c2f3744d59b848b449e4e54bb7/dcf8c6ec-33aa-4a2a-8bb2-c483f199f151.png)

*图 1.1：感知机模型示意图——输入信号经过加权求和后，通过激活函数产生输出*

#### 1.1.2 感知机的直观理解

可以将感知机想象为一个"投票表决系统"。每个输入 $x_i$ 代表一位投票者的意见（0 表示反对，1 表示赞成），权重 $w_i$ 代表该投票者的影响力大小。感知机将所有加权投票汇总后，如果总分超过某个阈值（由偏置 $b$ 控制），则输出"通过"（1），否则输出"不通过"（0）。

**示例：用感知机实现逻辑与（AND）运算**

逻辑与运算的真值表如下：

| $x_1$ | $x_2$ | AND 输出 |
|:---:|:---:|:---:|
| 0 | 0 | 0 |
| 0 | 1 | 0 |
| 1 | 0 | 0 |
| 1 | 1 | 1 |

我们可以设定 $w_1 = 0.5$，$w_2 = 0.5$，$b = -0.7$。验证：

- 输入 (0, 0)：$z = 0.5 \times 0 + 0.5 \times 0 - 0.7 = -0.7 < 0$，输出 0
- 输入 (0, 1)：$z = 0.5 \times 0 + 0.5 \times 1 - 0.7 = -0.2 < 0$，输出 0
- 输入 (1, 0)：$z = 0.5 \times 1 + 0.5 \times 0 - 0.7 = -0.2 < 0$，输出 0
- 输入 (1, 1)：$z = 0.5 \times 1 + 0.5 \times 1 - 0.7 = 0.3 \geq 0$，输出 1

结果完全符合逻辑与运算。类似地，逻辑或（OR）和逻辑非（NOT）也可以用单个感知机实现。

#### 1.1.3 感知机的局限性：异或问题

1969 年，Marvin Minsky 和 Seymour Papert 在其著作《Perceptrons》中证明了一个关键事实：**单层感知机无法解决异或（XOR）问题**。

异或运算的真值表如下：

| $x_1$ | $x_2$ | XOR 输出 |
|:---:|:---:|:---:|
| 0 | 0 | 0 |
| 0 | 1 | 1 |
| 1 | 0 | 1 |
| 1 | 1 | 0 |

从几何角度看，感知机本质上是在二维平面上画一条直线来分割两类数据。对于 AND 和 OR 运算，数据是**线性可分**的，一条直线就能将两类完美分开。但 XOR 的数据分布使得任何一条直线都无法将 (0,1)、(1,0) 与 (0,0)、(1,1) 分开——这就是**线性不可分**问题。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/cff72ea463dc4d3bb5fbb2b0fb7d4f0b/5bab73b6-feb3-47c1-8e83-9221d96b9c75.jpeg)

*线性可分与线性不可分示意图*

这一发现直接导致了人工智能的第一次"寒冬"。然而，解决方案其实很简单：**将多个感知机组合起来，形成多层网络**。

### 1.2 多层感知机与神经网络（Neural Network）

#### 1.2.1 从单层到多层

为了解决异或等线性不可分问题，研究者提出了**多层感知机（Multi-Layer Perceptron, MLP）**，也就是我们通常所说的**神经网络**。一个典型的神经网络由以下三种层组成：

| 层类型 | 英文名称 | 功能描述 |
|:---|:---|:---|
| **输入层**（Input Layer） | Input Layer | 接收原始数据，不进行任何计算，神经元数量等于输入特征的维度 |
| **隐藏层**（Hidden Layer） | Hidden Layer | 对数据进行非线性变换和特征提取，可以有一层或多层 |
| **输出层**（Output Layer） | Output Layer | 产生最终的预测结果，神经元数量取决于任务类型 |

隐藏层是神经网络的核心。每一层的每个神经元都执行与感知机相同的操作——加权求和加偏置，然后通过一个**激活函数**进行非线性变换。正是这种"多层叠加 + 非线性激活"的组合，赋予了神经网络拟合任意复杂函数的能力。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/264c68c2f3744d59b848b449e4e54bb7/571216a8-0db0-4527-99a4-4b83a0861cf4.png)

*图 1.2：多层神经网络结构示意图——包含输入层、隐藏层和输出层，层与层之间全连接*

面对上图中的多层神经网络结构，我们如何计算这个结构里面有多少参数？计算多层神经网络的参数量，我们需要理解参数的来源：权重（Weights）和偏置（Biases）。参数量是模型需要学习的总变量数，直接影响模型的复杂度和计算开销。

要计算多层神经网络的参数量，我们需要理解参数的来源：**权重（Weights）和偏置（Biases）**。参数量是模型需要学习的总变量数，直接影响模型的复杂度和计算开销。下面我们将以图1.2的全连接多层神经网络为案例，分步骤讲解计算方法。

**1. 理解网络结构**
首先，明确网络中每层的神经元数量（从左到右）：
- **输入层（蓝色）**：4个神经元（记为 $ n_{\text{input}} = 4 $）。
- **隐藏层1（绿色）**：5个神经元（记为 $ n_{\text{hidden1}} = 5 $）。
- **隐藏层2（棕色）**：3个神经元（记为 $ n_{\text{hidden2}} = 3 $）。
- **输出层（红色）**：2个神经元（记为 $ n_{\text{output}} = 2 $）。


**2. 参数计算的核心规则**
神经网络中，参数仅存在于**相邻两层的连接**中：
- **权重（Weights）**：前一层每个神经元与后一层每个神经元都有连接，因此权重数量 = **前一层神经元数 × 后一层神经元数**。
- **偏置（Biases）**：每个后层神经元都有一个偏置项，因此偏置数量 = **后一层神经元数**。


**3. 分层计算参数**
按照网络的连接顺序（输入层→隐藏层1→隐藏层2→输出层），逐段计算权重和偏置：

**（1）输入层 → 隐藏层1**
- 权重数量：$ n_{\text{input}} \times n_{\text{hidden1}} = 4 \times 5 = 20 $。
- 偏置数量：$ n_{\text{hidden1}} = 5 $。


**（2）隐藏层1 → 隐藏层2**
- 权重数量：$ n_{\text{hidden1}} \times n_{\text{hidden2}} = 5 \times 3 = 15 $。
- 偏置数量：$ n_{\text{hidden2}} = 3 $。


**（3）隐藏层2 → 输出层**
- 权重数量：$ n_{\text{hidden2}} \times n_{\text{output}} = 3 \times 2 = 6 $。
- 偏置数量：$ n_{\text{output}} = 2 $。


**4. 总参数量计算**
将所有权重和偏置相加：
- **总权重数**：$ 20 + 15 + 6 = 41 $。
- **总偏置数**：$ 5 + 3 + 2 = 10 $。
- **总参数量**：$ 41 + 10 = 51 $。


 **关键注意事项**
- **输入层无参数**：输入层仅接收数据，不参与参数学习（无权重和偏置）。
- **全连接层的假设**：图中是全连接网络（每个前层神经元连接后层所有神经元），若为稀疏连接（如卷积层），计算方法不同。
- **激活函数不影响参数**：参数计算与激活函数（如ReLU、Sigmoid）无关，仅与连接结构相关。

多层神经网络的参数量计算只需明确**每相邻两层的神经元数量**，按“权重=前×后，偏置=后”逐层计算，最后求和即可。掌握这一方法，可快速估算任意全连接网络的参数规模，为模型优化和资源分配提供依据。**注意以上计算方法仅适用于层与层之间全连接的情况。**


#### 1.2.2 神经网络的数学表示

以一个包含一个隐藏层的简单神经网络为例。假设输入层有 $d$ 个神经元，隐藏层有 $h$ 个神经元，输出层有 $q$ 个神经元。

**隐藏层的计算**：对于隐藏层的第 $j$ 个神经元（$j = 1, 2, \ldots, h$），其输出为：

$$
a_j = \sigma\left(\sum_{i=1}^{d} w_{ij}^{(1)} x_i + b_j^{(1)}\right)
$$

其中 $w_{ij}^{(1)}$ 是输入层第 $i$ 个神经元到隐藏层第 $j$ 个神经元的权重，$b_j^{(1)}$ 是隐藏层第 $j$ 个神经元的偏置，$\sigma$ 是激活函数。

用矩阵形式表示更加简洁：

$$
\mathbf{a} = \sigma(\mathbf{W}^{(1)} \mathbf{x} + \mathbf{b}^{(1)})
$$

**输出层的计算**：类似地，输出层的计算为：

$$
\mathbf{y} = g(\mathbf{W}^{(2)} \mathbf{a} + \mathbf{b}^{(2)})
$$

其中 $g$ 是输出层的激活函数（对于分类任务通常使用 Softmax，对于回归任务通常不使用激活函数或使用恒等函数）。

#### 1.2.3 万能近似定理

1989 年，George Cybenko 证明了**万能近似定理（Universal Approximation Theorem）**：只要隐藏层有足够多的神经元，一个包含单个隐藏层的前馈神经网络就能以任意精度逼近任何连续函数。这一定理为神经网络的强大表达能力提供了理论基础。

然而，"足够多的神经元"在实际中可能意味着天文数字。因此，现代深度学习倾向于使用**更多的层（更深的网络）**而非更宽的单层网络，因为深层网络可以通过逐层抽象来更高效地表示复杂函数。

### 1.3 解决异或问题：多层网络的威力

回到异或问题。我们可以构建一个包含 2 个输入神经元、2 个隐藏神经元和 1 个输出神经元的网络来解决它：

In [1]:
import numpy as np

# 异或问题的输入和期望输出
X = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
y_true = np.array([0, 1, 1, 0])

# 手动设定网络参数
# 隐藏层权重和偏置
W1 = np.array([[1, 1], [1, 1]])   # 2x2 权重矩阵
b1 = np.array([0, -1])            # 隐藏层偏置

# 输出层权重和偏置
W2 = np.array([1, -2])            # 1x2 权重向量
b2 = np.array([0])                # 输出层偏置

# 阶跃激活函数
def step(z):
    return (z >= 0).astype(int)

# 前向传播
for i in range(4):
    # 隐藏层计算
    z1 = X[i] @ W1 + b1            # 加权求和
    a1 = step(z1)                   # 激活
    # 输出层计算
    z2 = a1 @ W2 + b2
    y_pred = step(z2)
    print(f"输入: {X[i]}, 隐藏层输出: {a1}, 最终输出: {y_pred[0]}, 期望: {y_true[i]}")

输入: [0 0], 隐藏层输出: [1 0], 最终输出: 1, 期望: 0
输入: [0 1], 隐藏层输出: [1 1], 最终输出: 0, 期望: 1
输入: [1 0], 隐藏层输出: [1 1], 最终输出: 0, 期望: 1
输入: [1 1], 隐藏层输出: [1 1], 最终输出: 0, 期望: 0


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

多层网络通过隐藏层的非线性变换，将原本线性不可分的数据映射到了一个新的空间，在这个新空间中数据变得线性可分。

### 章节练习

**【单选题 1】** 单层感知机无法解决以下哪个逻辑运算问题？

A. 逻辑与（AND）
B. 逻辑或（OR）
C. 逻辑非（NOT）
D. 异或（XOR）

**【单选题 2】** 在一个三层神经网络（输入层-隐藏层-输出层）中，如果输入层有 4 个神经元，隐藏层有 6 个神经元，输出层有 3 个神经元，那么从输入层到隐藏层的权重矩阵的维度是？

A. 4 × 3
B. 6 × 4
C. 4 × 6
D. 3 × 6

**【简答题】** 请简要解释为什么单层感知机无法解决异或（XOR）问题，以及多层神经网络是如何克服这一局限性的。

## 第二章 神经网络的训练：前向传播与反向传播

### 2.1 数学预备知识：导数与偏导数

在学习神经网络的训练以及后续的梯度下降内容前，我们需要先掌握一些必要的数学基础——导数与偏导数。这些概念是理解梯度、反向传播和优化算法的数学基石。

#### 2.1.1 导数（Derivative）的概念

**导数**描述的是一个函数在某一点处的**变化率**（斜率）。直观地说，导数告诉我们：当输入发生一个微小变化时，输出会发生多大的变化。

对于一元函数 $f(x)$，它在点 $x_0$ 处的导数定义为：

$$
f'(x_0) = \lim_{\Delta x \to 0} \frac{f(x_0 + \Delta x) - f(x_0)}{\Delta x}
$$

导数的常见记法包括 $f'(x)$、$\frac{df}{dx}$、$\frac{dy}{dx}$。

**导数的几何意义**：导数等于函数图像在该点处切线的斜率。斜率为正表示函数在该点递增，斜率为负表示函数在该点递减，斜率为零表示该点可能是极值点。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/11d3bc42ee7446c797c7fa5f2a88df75/bece3a57-4436-4915-9620-31692f709e6b.png)

*图 ：导数的几何意义——函数在某点的导数等于该点切线的斜率*

#### 2.1.2 常见的求导规则

以下是深度学习中最常用的求导规则：

| 规则名称 | 公式 | 示例 |
|:---|:---|:---|
| **常数规则** | $\frac{d}{dx}c = 0$ | $\frac{d}{dx}5 = 0$ |
| **幂函数规则** | $\frac{d}{dx}x^n = nx^{n-1}$ | $\frac{d}{dx}x^3 = 3x^2$ |
| **常数倍数规则** | $\frac{d}{dx}[cf(x)] = c \cdot f'(x)$ | $\frac{d}{dx}3x^2 = 6x$ |
| **加法规则** | $\frac{d}{dx}[f(x)+g(x)] = f'(x)+g'(x)$ | $\frac{d}{dx}(x^2+3x) = 2x+3$ |
| **乘法规则** | $\frac{d}{dx}[f(x) \cdot g(x)] = f'g + fg'$ | $\frac{d}{dx}[x \cdot e^x] = e^x + xe^x$ |
| **链式法则** | $\frac{d}{dx}f(g(x)) = f'(g(x)) \cdot g'(x)$ | $\frac{d}{dx}e^{2x} = 2e^{2x}$ |
| **指数函数** | $\frac{d}{dx}e^x = e^x$ | $\frac{d}{dx}e^{-x} = -e^{-x}$ |
| **对数函数** | $\frac{d}{dx}\ln x = \frac{1}{x}$ | $\frac{d}{dx}\ln(2x) = \frac{1}{x}$ |

> **重点提示**：在神经网络中，**链式法则（Chain Rule）**是最重要的求导规则，因为神经网络本质上是多个函数的复合（嵌套），反向传播算法完全依赖链式法则来计算梯度。

**链式法则详解**：如果 $y = f(u)$ 且 $u = g(x)$，则：

$$
\frac{dy}{dx} = \frac{dy}{du} \cdot \frac{du}{dx}
$$

这可以推广到多层复合。例如，如果 $y = f(u)$、$u = g(v)$、$v = h(x)$，则：

$$
\frac{dy}{dx} = \frac{dy}{du} \cdot \frac{du}{dv} \cdot \frac{dv}{dx}
$$

**示例：求导 Sigmoid 函数**

Sigmoid 函数 $\sigma(x) = \frac{1}{1+e^{-x}}$ 在神经网络中广泛使用。我们来求它的导数：

$$
\sigma'(x) = \frac{d}{dx}\left(\frac{1}{1+e^{-x}}\right) = \frac{e^{-x}}{(1+e^{-x})^2} = \frac{1}{1+e^{-x}} \cdot \frac{e^{-x}}{1+e^{-x}} = \sigma(x) \cdot (1 - \sigma(x))
$$

这个优美的结果 $\sigma'(x) = \sigma(x)(1-\sigma(x))$ 意味着 Sigmoid 的导数可以用它自身的值来表示，这在反向传播中非常方便。

#### 2.1.3 偏导数（Partial Derivative）

当函数有**多个变量**时，我们需要用偏导数来描述函数对每个变量的变化率。偏导数的核心思想是：**固定其他变量不变，只对一个变量求导**。

对于二元函数 $f(x, y)$，它对 $x$ 的偏导数为：

$$
\frac{\partial f}{\partial x} = \lim_{\Delta x \to 0} \frac{f(x + \Delta x, y) - f(x, y)}{\Delta x}
$$

注意偏导数使用符号 $\partial$（而非 $d$），表示“偏”的含义。

**示例：均方误差损失函数的偏导数**

考虑一元线性回归的损失函数（均方误差）：

$$
L(w, b) = \frac{1}{m} \sum_{i=1}^{m} (wx_i + b - y_i)^2
$$

分别对 $w$ 和 $b$ 求偏导数：

$$
\frac{\partial L}{\partial w} = \frac{2}{m} \sum_{i=1}^{m} (wx_i + b - y_i) \cdot x_i
$$

$$
\frac{\partial L}{\partial b} = \frac{2}{m} \sum_{i=1}^{m} (wx_i + b - y_i)
$$

这两个偏导数告诉我们：当 $w$ 或 $b$ 发生微小变化时，损失函数 $L$ 会如何变化。梯度下降正是利用这些偏导数来更新参数的。


### 2.1 训练的核心目标

神经网络的训练本质上是一个**参数优化**过程。网络中有大量的权重 $w$ 和偏置 $b$ 需要确定，训练的目标是找到一组最优的参数值，使得网络在给定输入时能够产生尽可能接近真实标签的输出。

这个过程可以分为两个交替进行的阶段：

| 阶段 | 英文名称 | 核心任务 | 数据流方向 |
|:---|:---|:---|:---|
| **正向传播** | Forward Propagation | 根据当前参数计算预测输出 | 输入层 → 隐藏层 → 输出层 |
| **反向传播** | Back Propagation | 根据预测误差计算各参数的梯度 | 输出层 → 隐藏层 → 输入层 |

### 2.2 正向传播（Forward Propagation）

正向传播是数据从输入层经过隐藏层到达输出层的过程。在这个过程中，每一层的神经元接收上一层的输出，进行加权求和、加偏置、通过激活函数，然后将结果传递给下一层。

以一个简单的两层网络为例（一个隐藏层），正向传播的步骤如下：

**第一步：输入层到隐藏层**

$$
\mathbf{z}^{(1)} = \mathbf{W}^{(1)} \mathbf{x} + \mathbf{b}^{(1)}
$$
$$
\mathbf{a}^{(1)} = \sigma(\mathbf{z}^{(1)})
$$

**第二步：隐藏层到输出层**

$$
\mathbf{z}^{(2)} = \mathbf{W}^{(2)} \mathbf{a}^{(1)} + \mathbf{b}^{(2)}
$$
$$
\hat{\mathbf{y}} = g(\mathbf{z}^{(2)})
$$

**第三步：计算损失**

$$
L = \text{Loss}(\hat{\mathbf{y}}, \mathbf{y})
$$

其中 $\hat{\mathbf{y}}$ 是网络的预测输出，$\mathbf{y}$ 是真实标签，$L$ 是损失值。

### 2.3 反向传播（Back Propagation, BP）

反向传播算法由 David Rumelhart、Geoffrey Hinton 和 Ronald Williams 于 1986 年提出，是训练神经网络最核心的算法。它的核心思想是：**利用链式法则（Chain Rule），从输出层开始，逐层向后计算损失函数对每个参数的偏导数（梯度）**。

> **NOAI 考试重点**：2024 年和 2025 年的真题都考查了反向传播的作用。标准答案是：反向传播的主要作用是**调整网络的权重（Weight）和偏置项（Bias）**。

#### 2.3.1 链式法则回顾

链式法则是微积分中的基本法则。如果 $y = f(u)$，$u = g(x)$，则：

$$
\frac{dy}{dx} = \frac{dy}{du} \cdot \frac{du}{dx}
$$

对于多层网络，损失函数 $L$ 通过多层嵌套函数与某个权重 $w$ 相关联。链式法则允许我们将这个复杂的导数分解为一系列简单导数的乘积。

#### 2.3.2 反向传播的具体过程

继续使用上面的两层网络为例。假设使用均方误差损失函数 $L = \frac{1}{2}(\hat{y} - y)^2$，激活函数为 Sigmoid $\sigma(z) = \frac{1}{1+e^{-z}}$。

**第一步：计算输出层的误差信号**

$$
\delta^{(2)} = \frac{\partial L}{\partial \mathbf{z}^{(2)}} = (\hat{\mathbf{y}} - \mathbf{y}) \cdot g'(\mathbf{z}^{(2)})
$$

**第二步：将误差信号反向传播到隐藏层**

$$
\delta^{(1)} = (\mathbf{W}^{(2)})^T \delta^{(2)} \cdot \sigma'(\mathbf{z}^{(1)})
$$

**第三步：计算各层参数的梯度**

$$
\frac{\partial L}{\partial \mathbf{W}^{(2)}} = \delta^{(2)} (\mathbf{a}^{(1)})^T, \quad \frac{\partial L}{\partial \mathbf{b}^{(2)}} = \delta^{(2)}
$$

$$
\frac{\partial L}{\partial \mathbf{W}^{(1)}} = \delta^{(1)} \mathbf{x}^T, \quad \frac{\partial L}{\partial \mathbf{b}^{(1)}} = \delta^{(1)}
$$

**第四步：更新参数**

$$
\mathbf{W} \leftarrow \mathbf{W} - \alpha \frac{\partial L}{\partial \mathbf{W}}, \quad \mathbf{b} \leftarrow \mathbf{b} - \alpha \frac{\partial L}{\partial \mathbf{b}}
$$

其中 $\alpha$ 是学习率（Learning Rate）。

#### 2.3.3 反向传播的直觉理解

可以将反向传播理解为一个“责任追溯”系统。当网络的预测结果出现误差时，反向传播从输出层开始，逐层向前追溯每个神经元对这个误差的“贡献”。贡献大的神经元（即梯度大的参数）需要做更大幅度的调整，贡献小的则只需微调。这就像一个公司出了问题，需要从最终结果出发，逐级追查每个部门的责任，并据此进行整改。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/11d3bc42ee7446c797c7fa5f2a88df75/85a14934-3044-4c6e-aa0e-2b6fdea9a8ee.png)

*图 2.1：神经网络训练流程——正向传播计算预测值，反向传播计算梯度并更新参数*

### 2.4 训练过程的完整流程

一次完整的神经网络训练包含以下步骤，不断循环直到收敛：

```
1. 初始化：随机初始化所有权重和偏置
2. 循环（对于每个 epoch）：
   a. 正向传播：输入数据 → 逐层计算 → 得到预测输出
   b. 计算损失：比较预测输出与真实标签，计算损失值
   c. 反向传播：从输出层开始，逐层计算梯度
   d. 参数更新：使用梯度下降法更新所有权重和偏置
3. 终止条件：损失值足够小 / 达到最大迭代次数 / 验证集性能不再提升
```

### 章节练习

**【单选题 1】** 在训练神经网络时，反向传播算法（Back Propagation）的主要作用是什么？（此题改编自 NOAI 2024 真题第 13 题）

A. 计算网络的输出
B. 确定网络的结构
C. 评估网络的性能
D. 调整网络的权重和偏置项

**【单选题 2】** 在反向传播过程中，计算损失函数对各层参数的偏导数所依赖的数学工具是？

A. 泰勒展开（Taylor Expansion）
B. 链式法则（Chain Rule）
C. 贝叶斯定理（Bayes' Theorem）
D. 拉格朗日乘数法（Lagrange Multiplier）

**【简答题】** 请用自己的语言描述神经网络训练中正向传播和反向传播的过程，并说明两者的关系。（此题改编自 NOAI 2025 真题第 19 题）

## 第三章 梯度下降与优化方法

#### 3.1 梯度（Gradient）：偏导数的向量形式

将函数对所有变量的偏导数组合在一起，就得到了**梯度向量**：

$$
\nabla L(w, b) = \left(\frac{\partial L}{\partial w}, \frac{\partial L}{\partial b}\right)
$$

梯度向量指向函数值**增长最快**的方向。因此，沿着梯度的**反方向**移动，就能最快地减小函数值——这就是梯度下降的核心原理。

**导数、偏导数与梯度的关系总结**：

| 概念 | 适用场景 | 记号 | 直观含义 |
|:---|:---|:---|:---|
| **导数** | 一元函数 $f(x)$ | $f'(x)$ 或 $\frac{df}{dx}$ | 函数在该点的斜率（变化率） |
| **偏导数** | 多元函数 $f(x_1, x_2, \ldots)$ | $\frac{\partial f}{\partial x_i}$ | 固定其他变量，函数对某一变量的变化率 |
| **梯度** | 多元函数 $f(x_1, x_2, \ldots)$ | $\nabla f$ | 所有偏导数组成的向量，指向函数增长最快的方向 |

**Python 实践：数值求导**

在实际编程中，我们可以用数值方法近似计算导数：

In [3]:
import numpy as np

# 数值求导：使用中心差分法
def numerical_derivative(f, x, h=1e-5):
    return (f(x + h) - f(x - h)) / (2 * h)

# 示例：对 f(x) = x^2 求导，解析解为 f'(x) = 2x
f = lambda x: x ** 2  # 定义函数 f(x) = x^2，lambda 表示匿名函数, 用于快速定义简单的函数, 这里是一个一元函数, 输入为 x, 输出为 x^2, 返回 x^2, 即 x 的平方.
print(f"f'(3) 数值解 = {numerical_derivative(f, 3):.6f}")  # 约 6.0
print(f"f'(3) 解析解 = {2 * 3}")                           # 6

# 数值求偏导数
def numerical_partial_derivative(f, point, var_index, h=1e-5):
    """point 是一个列表，var_index 是要求偏导的变量索引"""
    point_plus = point.copy()
    point_minus = point.copy()
    point_plus[var_index] += h
    point_minus[var_index] -= h
    return (f(point_plus) - f(point_minus)) / (2 * h)

# 示例：对 f(x, y) = x^2 + 3xy + y^2 求偏导数
# 解析解：∂f/∂x = 2x + 3y，∂f/∂y = 3x + 2y
f2 = lambda p: p[0]**2 + 3*p[0]*p[1] + p[1]**2 # 定义函数 f(x, y) = x^2 + 3xy + y^2，lambda 表示匿名函数, 用于快速定义简单的函数。
# 这里是一个二元函数, 输入为一个包含两个元素的列表 p, 输出为 p[0]^2 + 3*p[0]*p[1] + p[1]^2, 返回 x^2 + 3xy + y^2, 即 x^2 + 3xy + y^2.
point = [2.0, 1.0]
print(f"\u2202f/\u2202x 在 点(2,1) = {numerical_partial_derivative(f2, point, 0):.4f}")  # 约 7.0
print(f"\u2202f/\u2202y 在 点(2,1) = {numerical_partial_derivative(f2, point, 1):.4f}")  # 约 8.0

f'(3) 数值解 = 6.000000
f'(3) 解析解 = 6
∂f/∂x 在 点(2,1) = 7.0000
∂f/∂y 在 点(2,1) = 8.0000


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

> **备考提示**：NOAI 第一轮不会要求复杂的求导计算，但你需要理解导数和偏导数的含义，能够读懂包含梯度计算的代码，并理解为什么梯度下降能够找到损失函数的最小值。

### 3.2 梯度下降（Gradient Descent）

梯度下降是机器学习和深度学习中最基础、最重要的优化算法。它的核心思想非常直观：**沿着损失函数下降最快的方向（即梯度的反方向）调整参数，逐步找到损失函数的最小值**。

> **NOAI 考试重点**：2024 年真题第 9 题直接考查了梯度下降法的主要目的——计算损失函数（Loss Function）的最小值。

#### 3.2.1 梯度的含义

对于一个多变量函数 $f(w_1, w_2, \ldots, w_n)$，其梯度是一个向量，由函数对每个变量的偏导数组成：

$$
\nabla f = \left(\frac{\partial f}{\partial w_1}, \frac{\partial f}{\partial w_2}, \ldots, \frac{\partial f}{\partial w_n}\right)
$$

梯度指向函数值增长最快的方向。因此，沿着梯度的**反方向**移动，就能最快地减小函数值。

#### 3.1.2 参数更新公式

梯度下降的参数更新公式为：

$$
\mathbf{w}_{t+1} = \mathbf{w}_t - \alpha \nabla L(\mathbf{w}_t)
$$

其中：
- $\mathbf{w}_t$ 是第 $t$ 步的参数值
- $\alpha$ 是**学习率（Learning Rate）**，控制每一步的步长大小
- $\nabla L(\mathbf{w}_t)$ 是损失函数在当前参数处的梯度

学习率的选择至关重要：学习率太大会导致参数在最优值附近来回震荡甚至发散；学习率太小则会导致收敛速度极慢。

#### 3.2.3 梯度下降的三种变体

根据每次更新时使用的数据量不同，梯度下降有三种主要变体：

| 变体 | 英文名称 | 每次更新使用的数据量 | 优点 | 缺点 |
|:---|:---|:---|:---|:---|
| 批量梯度下降 | Batch GD | 全部训练数据 | 梯度估计准确，收敛稳定 | 计算量大，内存需求高 |
| 随机梯度下降 | Stochastic GD (SGD) | 单个样本 | 计算快，能跳出局部最优 | 梯度估计噪声大，收敛不稳定 |
| 小批量梯度下降 | Mini-batch GD | 一小批样本（如 32、64） | 兼顾效率和稳定性 | 需要选择合适的批量大小 |

在实际深度学习训练中，**小批量梯度下降**是最常用的方法。

#### 3.2.4 梯度下降的 Python 实现

以下代码展示了用梯度下降法拟合一元线性回归 $y = wx + b$ 的过程（改编自 NOAI 2025 真题第 21 题）：

In [4]:
import numpy as np

# 生成模拟数据
np.random.seed(42)
x = np.random.rand(100) * 10
y = 2.5 * x + 3.0 + np.random.randn(100) * 2  # 真实参数: w=2.5, b=3.0

# 初始化参数
w = 0.0
b = 0.0
alpha = 0.001  # 学习率
m = len(x)     # 样本数量
epochs = 1000  # 迭代次数

# 梯度下降训练
for epoch in range(epochs):
    # 计算损失（均方误差）
    total_loss = 0
    dw = 0  # w 的梯度累加器
    db = 0  # b 的梯度累加器
    
    for i in range(m):
        y_pred = w * x[i] + b
        diff = y_pred - y[i]
        total_loss += diff ** 2
        dw += diff * x[i]   # 损失对 w 的偏导数
        db += diff           # 损失对 b 的偏导数
    
    # 平均梯度
    dw = dw / m
    db = db / m
    
    # 参数更新
    w = w - alpha * dw
    b = b - alpha * db
    
    if epoch % 200 == 0:
        print(f"Epoch {epoch}: Loss = {total_loss/(2*m):.4f}, w = {w:.4f}, b = {b:.4f}")

print(f"\n最终结果: w = {w:.4f}, b = {b:.4f}")
print(f"真实参数: w = 2.5000, b = 3.0000")

Epoch 0: Loss = 135.8318, w = 0.0905, b = 0.0148
Epoch 200: Loss = 2.7489, w = 2.8381, b = 0.6021
Epoch 400: Loss = 2.6293, w = 2.8191, b = 0.7554
Epoch 600: Loss = 2.5226, w = 2.7970, b = 0.8998
Epoch 800: Loss = 2.4271, w = 2.7760, b = 1.0363

最终结果: w = 2.7562, b = 1.1649
真实参数: w = 2.5000, b = 3.0000


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

**Epoch**（轮次）是指学习算法使用整个训练数据集完成一次前向传播和反向传播的过程。也就是模型把整个训练集完整“过”一遍的次数。

* 把数据集想象成一本书
* 1个Epoch = 把这本书从头到尾读一遍
* 读一遍不够？那就多读几遍（多轮迭代）

为什么需要多个Epoch？
Epoch 数量是训练前设定的超参数，其大小直接影响模型的收敛程度与泛化能力。模型不是天才，不可能看一眼就学会。每多一个Epoch，模型就多看一遍数据，参数（w、b）就更优化一步。

以日志行 为例：
`Epoch 0: Loss = 135.8318, w = 0.0905, b = 0.0148`  


| 字段 | 符号 | 解释 | 数值意义 |
|------|------|------|----------|
| **Epoch** | $t$ | 当前训练轮次 | 0 表示参数刚初始化，尚未更新 |
| **Loss** | $L$ | 损失函数在当前整个数据集（或代表性批次）上的平均值 | 量化模型预测值与真实值的差异，值越大说明模型性能越差 |
| **w** | $w$ | 线性模型中的权重（斜率） | 特征对输出的影响程度，初始时接近 0 |
| **b** | $b$ | 线性模型中的偏置（截距） | 不依赖特征的基准输出，初始时接近 0 |

---

**训练动态分析**

模型收敛（Convergence）指随着训练迭代（Epoch）的增加，模型的损失函数值逐渐下降并最终稳定在一个较小范围内波动，同时模型参数不再发生明显变化的状态。从上面代码的输出可观察到清晰的**收敛模式**：

- **Loss 持续下降**：  
(135.83 to 2.43)，下降约 98%，表明梯度下降法有效减少了经验风险。

- **参数逐步逼近最优值**：  
$w$从 0.09 增长至 2.78，$b$从 0.01 增长至 1.04，均向真实参数 $w=2.5, b=3.0$方向移动，说明优化方向正确。

- **收敛速度差异**：  
  权重 $w$ 已十分接近真实值（2.78 vs 2.5），而偏置 $b$ 仍有较大差距（1.04 vs 3.0）。  
  可能原因：
  - 数据分布中特征均值不为零，导致偏置的梯度较小；
  - 学习率设置对偏置的更新不足；
  - 训练轮次（Epoch）不够，偏置尚未充分收敛。



### 3.3 局部最优与全局最优

#### 3.3.1 概念定义

在优化问题中，存在两种"最优"：

**全局最优（Global Optima）**是整个参数空间中损失函数值最小的点。这是我们理想中希望找到的解。

**局部最优（Local Optima）**是某个局部区域内损失函数值最小的点，但不一定是全局最小。在这个点的附近，无论向哪个方向移动，损失函数值都会增大，但在更远的地方可能存在更低的点。

可以用一个形象的比喻来理解：想象你在一片山脉中寻找海拔最低的山谷。局部最优就像是某个小山谷的谷底——在这个山谷内它确实是最低点，但翻过旁边的山头可能还有更深的山谷。全局最优则是整片山脉中最深的那个山谷。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/be0e4b3afb0f4f9f949dc37ca6ce8160/1b9a03e5-1ef6-4ee8-a0f6-6e3dce6222b2.png)

*图 3.1：梯度下降与局部最优/全局最优示意图——梯度下降可能陷入局部最优而非找到全局最优*

#### 3.3.2 梯度下降与局部最优的关系

梯度下降法的一个固有缺陷是它可能会陷入局部最优。当参数到达一个局部最优点时，该点的梯度为零，参数更新量也为零，训练就会停滞。

在实际的深度学习中，由于网络参数空间的维度非常高（通常有数百万甚至数十亿个参数），纯粹的局部最优点其实很少见。更常见的问题是**鞍点（Saddle Point）**——在某些维度上是极小值，在另一些维度上是极大值的点。

#### 3.2.3 跳出局部最优的策略

| 策略 | 原理 | 实际做法 |
|:---|:---|:---|
| 随机初始化 | 不同的起点可能导向不同的最优解 | 多次随机初始化，选择最优结果 |
| 动量法 | 利用历史梯度的惯性冲过小的局部最优 | 使用 SGD with Momentum |
| 学习率调度 | 动态调整步长 | 使用学习率衰减或周期性学习率 |
| 随机性 | SGD 的噪声有助于跳出局部最优 | 使用小批量 SGD |

### 3.3 动量优化方法

#### 3.3.1 动量法（Momentum）

普通梯度下降只考虑当前的梯度方向，而动量法引入了"惯性"的概念——不仅考虑当前梯度，还考虑之前的更新方向。这就像一个球在山坡上滚动，即使遇到小的凸起（局部最优），由于惯性也能冲过去。

动量法的更新公式为：

$$
\mathbf{v}_t = \beta \mathbf{v}_{t-1} + \nabla L(\mathbf{w}_t)
$$
$$
\mathbf{w}_{t+1} = \mathbf{w}_t - \alpha \mathbf{v}_t
$$

其中 $\beta$ 是动量系数（通常取 0.9），$\mathbf{v}_t$ 是速度向量，它积累了历史梯度信息。

#### 3.3.2 Adam 优化器

**Adam（Adaptive Moment Estimation）**是目前深度学习中最常用的优化器之一，由 Diederik Kingma 和 Jimmy Ba 于 2014 年提出。Adam 结合了动量法和自适应学习率的思想，其核心创新在于：

1. **一阶矩估计（动量）**：跟踪梯度的指数移动平均，提供更新方向的惯性。
2. **二阶矩估计（自适应学习率）**：跟踪梯度平方的指数移动平均，为每个参数自动调整学习率。

Adam 的更新公式为：

$$
m_t = \beta_1 m_{t-1} + (1 - \beta_1) g_t \quad \text{（一阶矩估计）}
$$
$$
v_t = \beta_2 v_{t-1} + (1 - \beta_2) g_t^2 \quad \text{（二阶矩估计）}
$$
$$
\hat{m}_t = \frac{m_t}{1 - \beta_1^t}, \quad \hat{v}_t = \frac{v_t}{1 - \beta_2^t} \quad \text{（偏差修正）}
$$
$$
w_{t+1} = w_t - \frac{\alpha}{\sqrt{\hat{v}_t} + \epsilon} \hat{m}_t
$$

其中 $\beta_1 = 0.9$，$\beta_2 = 0.999$，$\epsilon = 10^{-8}$ 是默认超参数。

#### 3.3.3 AdamW 优化器

**AdamW** 是 Adam 的改进版本，由 Ilya Loshchilov 和 Frank Hutter 于 2017 年提出。它修正了 Adam 中权重衰减（Weight Decay）的实现方式。在标准 Adam 中，L2 正则化和权重衰减被混为一谈；而 AdamW 将权重衰减从梯度更新中解耦出来，直接在参数更新步骤中减去权重衰减项：

$$
w_{t+1} = w_t - \frac{\alpha}{\sqrt{\hat{v}_t} + \epsilon} \hat{m}_t - \alpha \lambda w_t
$$

其中 $\lambda$ 是权重衰减系数。AdamW 在训练大型模型（如 Transformer、大语言模型）时表现更好，已成为当前深度学习的默认优化器之一。

#### 3.3.4 Adam/AdamW 与普通梯度下降的对比

| 特性 | 普通 SGD | Adam / AdamW |
|:---|:---|:---|
| 学习率 | 所有参数共享同一个固定学习率 | 每个参数自适应调整学习率 |
| 动量 | 无（除非手动添加） | 内置一阶矩估计（动量） |
| 梯度缩放 | 无 | 通过二阶矩估计自动缩放梯度 |
| 超参数敏感性 | 对学习率非常敏感 | 默认超参数通常就能工作良好 |
| 收敛速度 | 较慢，需要仔细调参 | 通常更快收敛 |
| 适用场景 | 简单问题、需要精细调参时 | 大多数深度学习任务 |

### 章节练习

**【单选题 1】** 在机器学习中，梯度下降法（Gradient Descent Method）的主要目的是什么？（改编自 NOAI 2024 真题第 9 题）

A. 确定数据集的分布
B. 评估模型的复杂度
C. 计算损失函数的最小值
D. 计算损失函数的最大值

**【单选题 2】** 以下关于 Adam 优化器的描述，哪一项是正确的？

A. Adam 对所有参数使用相同的固定学习率
B. Adam 仅使用当前梯度进行参数更新，不考虑历史信息
C. Adam 结合了动量法和自适应学习率的思想
D. Adam 只适用于回归任务，不适用于分类任务

**【代码填空】** 以下代码使用梯度下降法训练一元线性回归模型。请补全标记处的代码。

```python
def gradient_descent(x, y, alpha=0.01, epochs=100):
    w, b = 0.0, 0.0
    m = len(x)
    
    for epoch in range(epochs):
        dw, db = 0.0, 0.0
        for i in range(m):
            y_pred = w * x[i] + b
            diff = y_pred - y[i]
            dw += ______(1)______   # w 的梯度累加
            db += ______(2)______   # b 的梯度累加
        
        dw /= m
        db /= m
        w = ______(3)______   # 更新 w
        b = ______(4)______   # 更新 b
    
    return w, b
```

## 第四章 卷积神经网络（CNN）

### 4.1 CNN 概述

**卷积神经网络（Convolutional Neural Network, CNN）**是一类专门用于处理具有网格结构数据（如图像）的深度学习模型。CNN 由 Yann LeCun 等人在 1990 年代提出，其经典模型 LeNet-5 被成功应用于手写数字识别。如今，CNN 已成为计算机视觉领域的基础架构。

> **NOAI 考试重点**：2024 年真题第 18-20 题以阅读材料的形式详细考查了卷积运算的数学定义和计算；2025 年真题第 7 题考查了 CNN 的典型层顺序，第 22 题考查了图像卷积的编程实现，第 23 题考查了用 PyTorch 构建 CNN。

CNN 的核心优势在于：

1. **局部连接**：每个神经元只与输入的一小块区域相连，而非全部，大大减少了参数数量。
2. **权值共享**：同一个卷积核在整个输入上滑动使用，进一步减少参数。
3. **平移不变性**：无论目标出现在图像的哪个位置，CNN 都能识别。

### 4.2 CNN 的网络结构

一个典型的 CNN 由以下几种层按顺序堆叠而成：

$$
\text{输入} \rightarrow [\text{卷积层} \rightarrow \text{激活函数} \rightarrow \text{池化层}] \times N \rightarrow \text{展平} \rightarrow \text{全连接层} \rightarrow \text{输出}
$$

> **NOAI 2025 真题第 7 题**考查了 CNN 前向传播的典型顺序。正确答案是：**卷积层 → 池化层 → 全连接层 → 输出层**。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/be0e4b3afb0f4f9f949dc37ca6ce8160/74bd118f-cdbd-417f-a83f-7f247b011fe5.png)

*图 4.1：CNN 典型架构示意图——输入图像经过卷积层、池化层、全连接层最终产生分类结果*

### 4.3 卷积层（Convolutional Layer）

#### 4.3.1 卷积运算的数学定义

在图像处理中，卷积运算的本质是：用一个小的矩阵（称为**卷积核**或**滤波器**，Kernel/Filter）在输入图像上滑动，在每个位置计算卷积核与对应区域的**逐元素乘积之和**。

> **NOAI 2024 真题**中将这种逐元素乘积之和称为**内积运算**（记作 $\odot$）。对于两个 $n \times n$ 的矩阵 $A$ 和 $B$，内积定义为：
>
> $$A \odot B = \sum_{i=1}^{n}\sum_{j=1}^{n} a_{ij} b_{ij}$$

对于一个 $m \times m$ 的卷积核 $A$ 和一个 $n \times n$ 的输入矩阵 $B$（$m \leq n$），卷积运算 $C = A * B$ 的结果是一个 $(n-m+1) \times (n-m+1)$ 的矩阵，其中：

$$
c_{i,j} = A \odot B_{sub}(i,j)
$$

这里 $B_{sub}(i,j)$ 是 $B$ 中从位置 $(i,j)$ 开始的 $m \times m$ 子矩阵。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/be0e4b3afb0f4f9f949dc37ca6ce8160/ce0cfda3-65f9-4ddc-99fb-d9b3f93b0292.png)

*图 4.2：卷积运算过程示意图——3×3 卷积核在 5×5 输入上滑动，逐元素相乘求和得到输出*

#### 4.3.2 卷积运算的计算示例

**示例 1：内积运算**（改编自 NOAI 2024 真题第 18 题）

已知 $A = \begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix}$，$B = \begin{pmatrix} 1 & 2 \\ 3 & 4 \end{pmatrix}$，则 $A \odot B$ 的计算过程为：

$$
A \odot B = 1 \times 1 + 1 \times 2 + 1 \times 3 + (-1) \times 4 = 1 + 2 + 3 - 4 = 2
$$

**示例 2：卷积运算**（改编自 NOAI 2024 真题第 19 题）

已知卷积核 $A = \begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix}$（$2 \times 2$），输入矩阵 $B = \begin{pmatrix} 1 & 1 & 1 \\ 1 & 2 & 1 \\ 1 & 1 & 1 \end{pmatrix}$（$3 \times 3$），则卷积结果 $C = A * B$ 是一个 $(3-2+1) \times (3-2+1) = 2 \times 2$ 的矩阵：

$$
c_{1,1} = A \odot \begin{pmatrix} 1 & 1 \\ 1 & 2 \end{pmatrix} = 1 \times 1 + 1 \times 1 + 1 \times 1 + (-1) \times 2 = 1
$$

$$
c_{1,2} = A \odot \begin{pmatrix} 1 & 1 \\ 2 & 1 \end{pmatrix} = 1 \times 1 + 1 \times 1 + 1 \times 2 + (-1) \times 1 = 3
$$

$$
c_{2,1} = A \odot \begin{pmatrix} 1 & 2 \\ 1 & 1 \end{pmatrix} = 1 \times 1 + 1 \times 2 + 1 \times 1 + (-1) \times 1 = 3
$$

$$
c_{2,2} = A \odot \begin{pmatrix} 2 & 1 \\ 1 & 1 \end{pmatrix} = 1 \times 2 + 1 \times 1 + 1 \times 1 + (-1) \times 1 = 3
$$

因此 $C = \begin{pmatrix} 1 & 3 \\ 3 & 3 \end{pmatrix}$。

#### 4.3.3 卷积层的关键参数

| 参数 | 英文名称 | 含义 | 对输出尺寸的影响 |
|:---|:---|:---|:---|
| 卷积核大小 | Kernel Size | 卷积核的宽和高，如 3×3、5×5 | 核越大，输出越小 |
| 步幅 | Stride | 卷积核每次滑动的像素数 | 步幅越大，输出越小 |
| 填充 | Padding | 在输入边缘补零的圈数 | 填充越多，输出越大 |
| 输出通道数 | Out Channels | 使用的卷积核数量 | 决定输出的通道数（深度） |

**输出尺寸计算公式**：

$$
\text{输出尺寸} = \left\lfloor \frac{\text{输入尺寸} - \text{卷积核大小} + 2 \times \text{填充}}{\text{步幅}} \right\rfloor + 1
$$

**示例**：输入图像大小为 $40 \times 40$，卷积核大小为 $3 \times 3$，步幅为 1，填充为 1：

$$
\text{输出尺寸} = \left\lfloor \frac{40 - 3 + 2 \times 1}{1} \right\rfloor + 1 = 40
$$

当 padding=1 且 kernel_size=3 时，输出尺寸与输入相同。

#### 4.3.4 常见卷积核的效果

不同的卷积核能提取图像的不同特征：

| 卷积核 | 名称 | 效果 |
|:---|:---|:---|
| $\begin{pmatrix} 1 & 1 & 1 \\ 1 & 1 & 1 \\ 1 & 1 & 1 \end{pmatrix} \times \frac{1}{9}$ | 均值模糊核 | 使图像变模糊（平滑） |
| $\begin{pmatrix} 0 & -1 & 0 \\ -1 & 5 & -1 \\ 0 & -1 & 0 \end{pmatrix}$ | 锐化核 | 增强图像边缘和细节 |
| $\begin{pmatrix} -1 & -1 & -1 \\ -1 & 8 & -1 \\ -1 & -1 & -1 \end{pmatrix}$ | 边缘检测核 | 提取图像中的边缘 |

> **NOAI 2024 真题第 20 题**给出了清华园图片的卷积前后对比，要求推断卷积核。卷积后图片呈现边缘效果，答案是边缘检测相关的卷积核。

### 4.4 池化层（Pooling Layer）

池化层的作用是**降低特征图的空间尺寸**，从而减少计算量和参数数量，同时增强特征的平移不变性。

#### 4.4.1 最大池化（Max Pooling）

在每个池化窗口中取最大值。例如，对一个 $4 \times 4$ 的特征图使用 $2 \times 2$ 的最大池化（步幅为 2）：

```
输入:                    输出:
| 1  3 | 2  4 |          | 5  7 |
| 5  2 | 7  1 |    →     | 5  6 |
|------|------|
| 3  1 | 4  2 |
| 5  0 | 3  6 |
```

每个 $2 \times 2$ 的区域取最大值：max(1,3,5,2)=5，max(2,4,7,1)=7，max(3,1,5,0)=5，max(4,2,3,6)=6。

> 注意：上面的示例中，左上角区域 max(1,3,5,2)=5，右上角 max(2,4,7,1)=7，左下角 max(3,1,5,0)=5，右下角 max(4,2,3,6)=6。输出为 $\begin{pmatrix} 5 & 7 \\ 5 & 6 \end{pmatrix}$。

#### 4.4.2 平均池化（Average Pooling）

在每个池化窗口中取平均值。使用同样的输入和 $2 \times 2$ 池化窗口：

左上角：(1+3+5+2)/4 = 2.75，右上角：(2+4+7+1)/4 = 3.5，以此类推。

#### 4.4.3 池化层的特点

池化层**没有可学习的参数**（不像卷积层有权重和偏置），它只是一个固定的下采样操作。池化后特征图的尺寸计算公式与卷积层相同。

### 4.5 全连接层（Fully Connected Layer）

全连接层（也称为**密集层**，Dense Layer）是传统神经网络中最基本的层类型。在 CNN 中，全连接层通常位于网络的末端，负责将卷积和池化层提取的特征映射到最终的输出类别。

在进入全连接层之前，需要将多维的特征图**展平（Flatten）**为一维向量。例如，一个大小为 $10 \times 10 \times 64$ 的特征图会被展平为一个长度为 6400 的向量。

在 PyTorch 中，展平操作通常使用 `x.view(x.size(0), -1)` 实现，其中 `x.size(0)` 是批量大小，`-1` 表示自动计算剩余维度。

### 4.6 CNN 完整结构示例

以 NOAI 2025 真题第 23 题中的验证码识别网络为例，其结构如下：

```
输入: 1×40×40 灰度图像
    ↓
第一个卷积块:
    Conv2d(1, 32, kernel_size=3, stride=1, padding=1)  → 32×40×40
    ReLU()
    MaxPool2d(kernel_size=2, stride=2)                  → 32×20×20
    ↓
第二个卷积块:
    Conv2d(32, 64, kernel_size=5, stride=1, padding=2)  → 64×20×20
    ReLU()
    MaxPool2d(kernel_size=2, stride=2)                  → 64×10×10
    ↓
展平: 64×10×10 = 6400
    ↓
全连接层:
    Linear(6400, 128)
    ReLU()
    Linear(128, 26)   → 输出 26 个类别（26 个字母）
```

### 章节练习

**【单选题 1】** 卷积神经网络（CNN）前向传播的典型层顺序是？（改编自 NOAI 2025 真题第 7 题）

A. 全连接层 → 卷积层 → 池化层 → 输出层
B. 池化层 → 卷积层 → 全连接层 → 输出层
C. 卷积层 → 池化层 → 全连接层 → 输出层
D. 卷积层 → 全连接层 → 池化层 → 输出层

**【单选题 2】** 已知 $A = \begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix}$，$B = \begin{pmatrix} 1 & 2 \\ 3 & 4 \end{pmatrix}$，则矩阵内积 $A \odot B$ 的值为？（改编自 NOAI 2024 真题第 18 题）

A. 10
B. 2
C. $\begin{pmatrix} 1 & 2 \\ 3 & -4 \end{pmatrix}$
D. $\begin{pmatrix} 4 & -2 \\ 6 & -2 \end{pmatrix}$

**【单选题 3】** 一个 $6 \times 6$ 的输入图像，使用 $3 \times 3$ 的卷积核，步幅为 1，无填充（padding=0），输出特征图的尺寸是？

A. $6 \times 6$
B. $5 \times 5$
C. $4 \times 4$
D. $3 \times 3$

**【代码填空】** 以下代码实现了一个简单的图像模糊函数（改编自 NOAI 2025 真题第 22 题）。对每个非边界像素，取其周围 8 个邻居像素值的平均值作为新值。请补全代码。

```python
import numpy as np

def blur(image, r=1):
    """对图像进行模糊处理，r 为模糊半径"""
    h, w = image.shape
    result = image.copy()
    
    for i in range(r, h - r):
        for j in range(r, w - r):
            tmp = 0
            count = 0
            for di in range(-r, r + 1):
                for dj in range(-r, r + 1):
                    if di == 0 and dj == 0:
                        continue
                    tmp += ______(1)______   # 累加邻居像素值
                    count += 1
            result[i][j] = ______(2)______   # 计算平均值
    
    return result
```

---

## 第五章 决策函数、激活函数与损失函数

> **NOAI 考试重点**：2025 年真题第 8 题以匹配题的形式考查了三种函数的区别。理解这三种函数各自的作用是 NOAI 考试的核心考点。

### 5.1 三种函数的角色定位

在神经网络中，有三种至关重要的函数，它们各自承担不同的角色：

| 函数类型 | 英文名称 | 核心作用 | 在网络中的位置 |
|:---|:---|:---|:---|
| **决策函数** | Decision Function | 将模型输出转化为最终的预测结果 | 网络的最后一步 |
| **激活函数** | Activation Function | 引入非线性，增强网络的表达能力 | 每一层的输出 |
| **损失函数** | Loss Function | 衡量预测结果与真实标签之间的差距 | 训练过程中计算误差 |

### 5.2 激活函数（Activation Function）

> **NOAI 2024 真题第 14 题**：在神经网络中，将线性求和的结果进行非线性映射的是**激活函数**。

激活函数的作用是**为神经网络引入非线性**。如果没有激活函数，无论网络有多少层，整个网络都只是一个线性变换，等价于单层网络。

#### 5.2.1 Sigmoid 函数

$$
\sigma(x) = \frac{1}{1 + e^{-x}}
$$

Sigmoid 将任意实数映射到 $(0, 1)$ 区间，输出可以解释为概率。它的导数为 $\sigma'(x) = \sigma(x)(1 - \sigma(x))$。

**优点**：输出范围有界，适合表示概率。**缺点**：存在梯度消失问题——当输入值很大或很小时，导数接近于零，导致深层网络的梯度几乎无法传播。

#### 5.2.2 ReLU 函数

$$
\text{ReLU}(x) = \max(0, x) = \begin{cases} x, & \text{if } x > 0 \\ 0, & \text{if } x \leq 0 \end{cases}
$$

> **NOAI 2025 真题第 8 题**中对 ReLU 的描述是：**引入非线性，输出为输入与零的较大值**。

ReLU（Rectified Linear Unit）是目前最常用的激活函数。它的计算非常简单，且在正区间内梯度恒为 1，有效缓解了梯度消失问题。

**优点**：计算简单高效，缓解梯度消失。**缺点**：存在"神经元死亡"问题——如果某个神经元的输入始终为负，其输出永远为零，梯度也永远为零，该神经元将永远无法被激活。

#### 5.2.3 Tanh 函数

$$
\tanh(x) = \frac{e^x - e^{-x}}{e^x + e^{-x}}
$$

Tanh 将输入映射到 $(-1, 1)$ 区间，输出以零为中心。它可以看作是 Sigmoid 的缩放和平移版本：$\tanh(x) = 2\sigma(2x) - 1$。

#### 5.2.4 激活函数对比

| 激活函数 | 输出范围 | 优点 | 缺点 | 常用场景 |
|:---|:---|:---|:---|:---|
| Sigmoid | $(0, 1)$ | 输出可解释为概率 | 梯度消失，非零中心 | 二分类输出层 |
| Tanh | $(-1, 1)$ | 零中心输出 | 梯度消失 | RNN 隐藏层 |
| ReLU | $[0, +\infty)$ | 计算快，缓解梯度消失 | 神经元死亡 | CNN 和 MLP 隐藏层 |

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/be0e4b3afb0f4f9f949dc37ca6ce8160/a9e1d68c-db3d-405d-9931-5b3d8ab09225.png)

*图 5.1：三种常见激活函数的图像对比——Sigmoid、Tanh 和 ReLU 的输出范围和形状差异*

### 5.3 决策函数（Decision Function）

决策函数位于网络的最后一步，负责将模型的原始输出（logits）转化为可解释的预测结果。

#### 5.3.1 Softmax 函数

> **NOAI 2025 真题第 8 题**中对 Softmax 的描述是：**将多个输出值转化为概率分布，用于分类任务**。

Softmax 函数将一个 $K$ 维的实数向量转化为一个概率分布，使得所有输出值都在 $(0, 1)$ 之间且总和为 1：

$$
\text{Softmax}(z_i) = \frac{e^{z_i}}{\sum_{j=1}^{K} e^{z_j}}
$$

**计算示例**（改编自 NOAI 2025 真题第 11 题）：

假设一个三分类模型的原始输出为 $[2.0, 1.0, 1.0]$（分别对应猫、狗、鸟），则 Softmax 计算过程为：

$$
e^{2.0} = 7.389, \quad e^{1.0} = 2.718, \quad e^{1.0} = 2.718
$$

$$
\text{总和} = 7.389 + 2.718 + 2.718 = 12.825
$$

$$
P(\text{猫}) = \frac{7.389}{12.825} \approx 0.576, \quad P(\text{狗}) = \frac{2.718}{12.825} \approx 0.212, \quad P(\text{鸟}) = \frac{2.718}{12.825} \approx 0.212
$$

最终预测为"猫"（概率最高的类别）。

#### 5.3.2 Sigmoid 作为决策函数

在二分类问题中，Sigmoid 函数也可以作为决策函数使用。输出值大于 0.5 则预测为正类，否则预测为负类。

### 5.4 损失函数（Loss Function）

> **NOAI 2025 真题第 8 题**中对交叉熵损失函数的描述是：**计算预测结果与真实标签之间的误差，指导网络优化**。

损失函数衡量模型预测值与真实值之间的差距，是训练过程中优化的目标。

#### 5.4.1 均方误差（MSE）

$$
L_{MSE} = \frac{1}{n}\sum_{i=1}^{n}(\hat{y}_i - y_i)^2
$$

适用于**回归任务**。它计算预测值与真实值之差的平方的平均值。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/be0e4b3afb0f4f9f949dc37ca6ce8160/dbd4ba6a-5c18-4f1e-b155-d4edf7595208.png)

*图 5.2：均方误差（MSE）与交叉熵损失函数的对比——两者分别适用于回归和分类任务*

#### 5.4.2 交叉熵损失（Cross-Entropy Loss）

**二分类交叉熵**：

$$
L_{BCE} = -\frac{1}{n}\sum_{i=1}^{n}[y_i \log(\hat{y}_i) + (1-y_i)\log(1-\hat{y}_i)]
$$

**多分类交叉熵**：

$$
L_{CE} = -\frac{1}{n}\sum_{i=1}^{n}\sum_{k=1}^{K} y_{ik} \log(\hat{y}_{ik})
$$

交叉熵损失适用于**分类任务**，通常与 Softmax 决策函数配合使用。

#### 5.4.3 损失函数对比

| 损失函数 | 适用任务 | 输出层激活函数 | 特点 |
|:---|:---|:---|:---|
| 均方误差（MSE） | 回归 | 无（恒等函数） | 对异常值敏感 |
| 二分类交叉熵（BCE） | 二分类 | Sigmoid | 梯度特性好 |
| 多分类交叉熵（CE） | 多分类 | Softmax | 最常用的分类损失 |

### 章节练习

**【单选题 1】** 以下关于 Softmax、ReLU 和交叉熵损失函数的匹配，哪一项是正确的？（改编自 NOAI 2025 真题第 8 题）

A. Softmax → 激活函数，ReLU → 决策函数，交叉熵 → 损失函数
B. Softmax → 损失函数，ReLU → 激活函数，交叉熵 → 决策函数
C. Softmax → 决策函数，ReLU → 激活函数，交叉熵 → 损失函数
D. Softmax → 决策函数，ReLU → 损失函数，交叉熵 → 激活函数

**【单选题 2】** 在神经网络中，以下哪个概念指的是将线性求和的结果进行非线性映射？（改编自 NOAI 2024 真题第 14 题）

A. 权重（Weight）
B. 激活函数（Activation Function）
C. 偏置项（Intercept）
D. 层间连接（Connection）

**【代码填空】** 以下代码实现了 Softmax 函数。请补全标记处的代码。

```python
import numpy as np

def softmax(z):
    """计算 Softmax 概率分布"""
    exp_z = ______(1)______        # 计算每个元素的指数
    sum_exp = ______(2)______      # 计算指数之和
    return exp_z / sum_exp

# 测试
logits = np.array([2.0, 1.0, 1.0])
probs = softmax(logits)
print(f"概率分布: {probs}")
print(f"预测类别: {np.argmax(probs)}")
```

## 第六章 大语言模型（LLM）核心思想

### 6.1 从神经网络到大语言模型

大语言模型（LLM）是一类基于海量文本语料、使用深度学习（以Transformer架构为主）训练的大规模神经网络，其核心目标是建模自然语言的概率分布，通常通过“预测下一个词”的自监督任务学习通用的语言知识与世界知识。LLM的特点在于参数规模巨大（通常达数十亿至数千亿）、训练数据极广，因而具备强大的语言理解、生成与上下文学习能力。代表性模型如GPT、BERT、Llama、文心一言等，已成为生成式人工智能的基础设施。大语言模型（Large Language Model, LLM）是近年来人工智能领域最重要的突破之一。从 GPT 系列到 Claude、Gemini，这些模型展现出了惊人的语言理解和生成能力。

### 6.2 Transformer 架构

2017 年，Google 团队在论文《Attention Is All You Need》中提出了 **Transformer** 架构，这是所有现代大语言模型的基础。Transformer 的核心创新是**自注意力机制（Self-Attention）**，它允许模型在处理序列中的每个位置时，同时关注序列中所有其他位置的信息。

与此前的循环神经网络（RNN）逐步处理序列不同，Transformer 可以**并行处理**整个序列，这大大提高了训练效率，使得训练超大规模模型成为可能。

#### 6.2.1 自注意力机制

自注意力机制的核心思想可以用一个直观的比喻来理解：当你阅读一个句子时，理解某个词的含义往往需要参考句子中的其他词。例如，在"小明把球传给了小红，**她**很高兴"这句话中，理解"她"指的是谁，需要关注前文中的"小红"。

自注意力机制通过三个矩阵来实现这种"关注"：

| 矩阵 | 英文名称 | 作用 |
|:---|:---|:---|
| **查询矩阵** Q | Query | "我想找什么信息？" |
| **键矩阵** K | Key | "我有什么信息可以提供？" |
| **值矩阵** V | Value | "我实际要传递的信息是什么？" |

注意力分数的计算公式为：

$$
\text{Attention}(Q, K, V) = \text{Softmax}\left(\frac{QK^T}{\sqrt{d_k}}\right)V
$$

其中 $d_k$ 是键向量的维度，除以 $\sqrt{d_k}$ 是为了防止点积值过大导致 Softmax 输出过于极端。

### 6.3 LLM 的核心思想：下一个词预测

大语言模型的核心训练目标出奇地简单：**给定一段文本的前文，预测下一个最可能出现的词（token）**。这被称为**自回归语言建模（Autoregressive Language Modeling）**。

例如，给定前文"今天天气真"，模型需要预测下一个词最可能是"好"、"热"、"冷"等。通过在海量文本数据上反复训练这个简单的任务，模型逐渐学会了语言的语法、语义、逻辑推理，甚至世界知识。

这个过程可以形式化为：给定词序列 $w_1, w_2, \ldots, w_{t-1}$，模型输出下一个词 $w_t$ 的概率分布：

$$
P(w_t | w_1, w_2, \ldots, w_{t-1})
$$

在生成文本时，模型从这个概率分布中采样一个词，将其添加到序列末尾，然后重复这个过程，逐词生成完整的文本。

### 6.4 LLM 的关键特征

#### 6.4.1 规模效应（Scaling Law）

研究发现，LLM 的能力随着三个因素的增长而持续提升：

1. **模型参数量**：更多的参数意味着更强的表达能力
2. **训练数据量**：更多的数据让模型学到更丰富的知识
3. **计算量**：更多的计算资源允许更充分的训练

当这三个因素同时增大到一定程度时，模型会出现**涌现能力（Emergent Abilities）**——即在小模型上完全不具备、但在大模型上突然出现的能力，如复杂推理、代码生成等。

#### 6.4.2 预训练与微调

LLM 的训练通常分为两个阶段：

**预训练（Pre-training）**：在海量无标注文本数据上进行自回归语言建模训练。这个阶段让模型学会语言的基本规律和广泛的世界知识。预训练需要巨大的计算资源（数千张 GPU 训练数周甚至数月）。

**微调（Fine-tuning）**：在特定任务的标注数据上进一步训练，使模型适应具体的应用场景。常见的微调方法包括监督微调（SFT）和基于人类反馈的强化学习（RLHF）。

#### 6.4.3 Tokenization

LLM 不直接处理文字，而是先将文本切分为**token**（词元）。Token 可以是一个完整的词、一个子词，甚至是一个字符。例如，"unbelievable" 可能被切分为 "un"、"believ"、"able" 三个 token。

常用的 tokenization 方法包括 BPE（Byte Pair Encoding）和 WordPiece。中文文本通常按字或词进行切分。

#### 6.4.4 提示工程（Prompt Engineering）

> **提示工程（Prompt Engineering）** 是指通过**系统化设计输入提示（prompt）**，引导大型语言模型（LLM）生成符合特定任务目标、格式要求和语义约束的输出的**方法论**。其本质是**利用模型的预训练知识**，通过**输入结构设计**实现**零样本（Zero-shot）或少样本（Few-shot）学习**，避免对模型进行参数微调（Fine-tuning）。

> **权威来源**：  
> Brown et al. (2020) 在《Language Models are Few-Shot Learners》中首次系统提出：  
> *"Prompt engineering is the process of constructing inputs to language models that elicit desired behaviors without modifying the model weights."*  
> （提示工程是构建输入以激发模型行为的流程，无需修改模型权重。）

在撰写提示词时，我们可以参考以下原则：                                     
**明确指令（Avoid Ambiguity）**❌ 低效提示："写点关于AI的内容"
✅ 有效提示："用200字简述AI在医疗领域的3个应用，用列表形式，语言简洁易懂"
💡 原理：AI无法猜你的意图，需要明确任务细节。例如明确输出文本的长度，指定输出内容的格式和避免否定内容等。

**提供上下文（Context is King）**
❌ 低效提示："翻译成英语"
✅ 有效提示："将以下中文医疗报告翻译成专业医学英语：[报告内容]。要求使用WHO标准术语，避免口语化"
💡 原理：AI需要背景才能输出精准内容。

**示例示范（Few-shot Learning）**
❌ 低效提示："写一首关于春天的诗"
✅ 有效提示：
"请模仿以下示例写一首关于春天的诗：
示例1：'细雨轻风拂柳枝，新芽破土见春晖。'
示例2：'桃红柳绿莺声脆，燕子归来筑巢忙。'
要求：押韵，5言绝句，描写自然景象"
💡 原理：通过示例告诉AI"你想要什么风格"。

**角色设定（Role Play）**
❌ 低效提示："解释量子力学"
✅ 有效提示："你是一位初中物理老师，请用比喻解释量子纠缠，避免数学公式，用'水波纹'类比"
💡 原理：设定角色让AI调整语言风格和深度。

**思维链（Chain-of-Thought）** ❌ 低效提示："计算1+1"✅ 有效提示："逐步解释：1个苹果加1个苹果等于几个苹果？为什么？"
💡 原理：让AI"展示思考过程"，大幅提高复杂问题的准确性。

提示工程的本质不是让AI变聪明，而是让你的提问更聪明！
好的提示 = 清晰的指令 + 适当的上下文 + 明确的输出要求
✨ 记住：
"AI不是魔法，而是你和它的对话方式。
你问得清楚，它就答得明白。"

### 6.5 LLM 的局限性

尽管 LLM 能力强大，但它们也存在明显的局限性：

| 局限性 | 描述 |
|:---|:---|
| **幻觉（Hallucination）** | 模型可能生成看似合理但实际上不正确的内容 |
| **知识截止** | 模型的知识仅限于训练数据的时间范围 |
| **推理能力有限** | 在复杂的逻辑推理和数学计算上仍有不足 |
| **上下文长度限制** | 能处理的文本长度有上限 |

### 章节练习

**【单选题 1】** 大语言模型（LLM）的核心训练目标是什么？

A. 将输入文本翻译成另一种语言
B. 给定前文，预测下一个最可能出现的词
C. 对输入文本进行情感分类
D. 从文本中提取命名实体

**【单选题 2】** Transformer 架构的核心创新是什么？

A. 循环连接（Recurrent Connection）
B. 卷积运算（Convolution）
C. 自注意力机制（Self-Attention）
D. 池化操作（Pooling）

**【简答题】** 请简要描述大语言模型（LLM）的核心思想，并解释什么是"涌现能力"。

## 第七章 Python 面向对象编程基础

在学习 PyTorch 之前，我们需要先掌握 Python 的**面向对象编程（Object-Oriented Programming, OOP）**基础。PyTorch 中的所有神经网络模型都是通过类的继承来定义的，因此理解类与对象是读懂和编写 PyTorch 代码的必要前提。

![alt](https://bohrium.oss-cn-zhangjiakou.aliyuncs.com/article/1837017/be0e4b3afb0f4f9f949dc37ca6ce8160/0baf9d0a-54dc-44c4-a607-51e16af1df90.png)

*图 7.0：类与对象的关系示意图——类是蓝图，对象是根据蓝图建造的实体*

### 7.1 类（Class）与对象（Object）的基本概念

**类**是一种“蓝图”或“模板”，它定义了一类事物共有的属性（数据）和方法（行为）。**对象**是根据类创建的具体实例。举个类比：“狗”这个概念是一个类，而你家的那只叫“旺财”的金毛犬就是一个对象。

#### 7.1.1 定义一个类

In [5]:
class Dog:
    """狗类——定义了狗的属性和行为"""
    
    # __init__ 是构造方法（初始化方法），创建对象时自动调用
    def __init__(self, name, breed, age):
        self.name = name      # self.name 是实例属性
        self.breed = breed
        self.age = age
    
    # 实例方法：定义对象的行为
    def bark(self):
        return f"{self.name} says: Woof!"
    
    def info(self):
        return f"{self.name} is a {self.age}-year-old {self.breed}"

# 创建对象（实例化）
my_dog = Dog("Buddy", "Golden Retriever", 3)
print(my_dog.bark())   # 输出: Buddy says: Woof!
print(my_dog.info())   # 输出: Buddy is a 3-year-old Golden Retriever
print(my_dog.name)     # 输出: Buddy（访问属性）

Buddy says: Woof!
Buddy is a 3-year-old Golden Retriever
Buddy


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

**关键概念解析**：

| 概念 | 说明 | 示例 |
|:---|:---|:---|
| `class` | 定义类的关键字 | `class Dog:` |
| `__init__` | 构造方法，创建对象时自动调用 | `def __init__(self, name):` |
| `self` | 代表对象自身，用于访问属性和方法 | `self.name = name` |
| 实例属性 | 属于对象的数据 | `self.name`、`self.age` |
| 实例方法 | 属于对象的函数 | `def bark(self):` |
| 实例化 | 根据类创建对象 | `my_dog = Dog("Buddy", ...)` |

#### 7.1.2 类的继承（Inheritance）

继承是 OOP 的核心特性之一，它允许一个类（子类）继承另一个类（父类）的属性和方法，并可以添加新的功能或修改已有功能。**这在 PyTorch 中至关重要**，因为所有自定义的神经网络模型都必须继承 `nn.Module` 类。

In [6]:
# 父类：动物
class Animal:
    def __init__(self, name, species):
        self.name = name
        self.species = species
    
    def speak(self):
        return f"{self.name} makes a sound"

# 子类：狗（继承自 Animal）
class Dog(Animal):
    def __init__(self, name, breed):
        # 调用父类的 __init__ 方法
        super().__init__(name, species="Canine")
        self.breed = breed  # 子类新增的属性
    
    # 重写父类的方法
    def speak(self):
        return f"{self.name} says: Woof!"
    
    # 子类新增的方法
    def fetch(self, item):
        return f"{self.name} fetches the {item}"

# 子类：猫（继承自 Animal）
class Cat(Animal):
    def __init__(self, name):
        super().__init__(name, species="Feline")
    
    def speak(self):
        return f"{self.name} says: Meow!"

# 使用
dog = Dog("Buddy", "Labrador")
cat = Cat("Whiskers")

print(dog.speak())         # Buddy says: Woof!  （调用子类重写的方法）
print(dog.species)         # Canine             （继承自父类的属性）
print(dog.fetch("ball"))   # Buddy fetches the ball  （子类新增的方法）
print(cat.speak())         # Whiskers says: Meow!
print(isinstance(dog, Animal))  # True（dog 也是 Animal 类型）

Buddy says: Woof!
Canine
Buddy fetches the ball
Whiskers says: Meow!
True


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

**继承的关键要点**：

| 概念 | 说明 | PyTorch 中的对应 |
|:---|:---|:---|
| `super()` | 调用父类的方法 | `super().__init__()` 初始化 nn.Module |
| 方法重写 | 子类重新定义父类的方法 | 重写 `forward()` 方法定义前向传播 |
| 新增属性/方法 | 子类可以扩展父类 | 添加卷积层、全连接层等 |

### 7.2 为什么 PyTorch 需要类与对象

在 PyTorch 中，定义一个神经网络模型的标准方式是创建一个继承自 `nn.Module` 的类。下面是一个直观的对比：

In [8]:
pip install torch

In [9]:
import torch
import torch.nn as nn

# PyTorch 中定义神经网络的标准方式
class SimpleNet(nn.Module):          # 继承 nn.Module
    def __init__(self, input_size, hidden_size, output_size):
        super().__init__()           # 调用父类构造方法
        # 定义网络层（实例属性）
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, output_size)
    
    def forward(self, x):            # 重写 forward 方法
        x = self.fc1(x)              # 第一层全连接
        x = self.relu(x)             # 激活函数
        x = self.fc2(x)              # 第二层全连接
        return x

# 创建模型对象（实例化）
model = SimpleNet(input_size=784, hidden_size=128, output_size=10)

# 使用模型进行预测（调用对象的方法）
fake_input = torch.randn(1, 784)
output = model(fake_input)  # 自动调用 forward 方法
print(output.shape)         # torch.Size([1, 10])

torch.Size([1, 10])


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

可以看到，PyTorch 的模型定义完全基于类的继承机制：

1. **继承 `nn.Module`**：获得参数管理、GPU 迁移、模型保存/加载等能力
2. **`__init__` 中定义层**：用 `self.xxx` 保存网络层作为实例属性
3. **重写 `forward` 方法**：定义数据如何在网络中流动（前向传播）

### 7.3 特殊方法（Magic Methods）

Python 类中以双下划线开头和结尾的方法被称为特殊方法（也叫魔术方法），它们允许我们自定义对象的行为：

In [10]:
class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y
    
    def __repr__(self):
        """__repr__: 定义对象的字符串表示"""
        return f"Vector({self.x}, {self.y})"
    
    def __add__(self, other):
        """__add__: 定义 + 运算符的行为"""
        return Vector(self.x + other.x, self.y + other.y)
    
    def __len__(self):
        """__len__: 定义 len() 的行为"""
        return 2  # 二维向量
    
    def __call__(self, scalar):
        """__call__: 让对象可以像函数一样被调用"""
        return Vector(self.x * scalar, self.y * scalar)

v1 = Vector(1, 2)
v2 = Vector(3, 4)
print(v1)           # Vector(1, 2)      ← 触发 __repr__
print(v1 + v2)      # Vector(4, 6)      ← 触发 __add__
print(len(v1))      # 2                 ← 触发 __len__
print(v1(3))        # Vector(3, 6)      ← 触发 __call__

Vector(1, 2)
Vector(4, 6)
2
Vector(3, 6)


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

> **与 PyTorch 的关联**：当你写 `output = model(input)` 时，实际上触发的是 `nn.Module` 的 `__call__` 方法，它内部会调用你定义的 `forward` 方法。这就是为什么我们只需要重写 `forward` 而不是 `__call__`。

### 章节练习

**【单选题 1】** 在 Python 中，以下哪个方法在创建对象时会自动被调用？

A. `__call__`
B. `__init__`
C. `__repr__`
D. `forward`

**【单选题 2】** 在 PyTorch 中定义神经网络时，以下哪个说法是正确的？

A. 自定义模型类必须继承 `nn.Module` 并重写 `forward` 方法
B. 自定义模型类不需要调用 `super().__init__()`
C. 网络层应该在 `forward` 方法中定义，而不是 `__init__` 中
D. 调用 `model(input)` 时会直接执行 `__init__` 方法

**【代码填空题】** 请填写以下代码中的空白处，完成一个简单的 PyTorch 模型定义：

```python
import torch.nn as nn

class MyModel(__(1)__):       # 继承父类
    def __init__(self):
        __(2)__                # 调用父类构造方法
        self.layer1 = nn.Linear(10, 5)
        self.relu = nn.ReLU()
        self.layer2 = nn.Linear(5, 2)
    
    def __(3)__(self, x):     # 定义前向传播
        x = self.layer1(x)
        x = self.relu(x)
        x = self.layer2(x)
        return x
```

## 第八章 PyTorch 框架基础

### 8.1 PyTorch 概述

**PyTorch** 是由 Meta（原 Facebook）AI 研究院开发的开源深度学习框架，是当前学术界和工业界最流行的深度学习工具之一。PyTorch 以其**动态计算图**和 **Pythonic** 的编程风格著称，使得模型的构建和调试都非常直观。

> **NOAI 考试重点**：NOAI 第二轮要求参与者掌握 PyTorch 框架的基本应用，第一轮笔试中也会出现 PyTorch 相关的代码填空题（如 2025 年真题第 23 题）。

### 8.2 张量（Tensor）基础

张量是 PyTorch 中最基本的数据结构，可以理解为多维数组。它与 NumPy 的 ndarray 非常相似，但张量可以在 GPU 上运算，并且支持自动微分。

#### 8.2.1 张量的创建

In [11]:
import torch

# 从 Python 列表创建张量
a = torch.tensor([1, 2, 3])
print(a)          # tensor([1, 2, 3])
print(a.shape)    # torch.Size([3])
print(a.dtype)    # torch.int64

# 创建浮点数张量
b = torch.tensor([1.0, 2.0, 3.0])
print(b.dtype)    # torch.float32

# 创建多维张量（矩阵）
c = torch.tensor([[1, 2, 3], [4, 5, 6]])
print(c.shape)    # torch.Size([2, 3])

# 创建特殊张量
zeros = torch.zeros(3, 4)       # 3×4 全零张量
ones = torch.ones(2, 3)         # 2×3 全一张量
rand = torch.rand(2, 3)         # 2×3 均匀分布随机张量 [0, 1)
randn = torch.randn(2, 3)       # 2×3 标准正态分布随机张量
eye = torch.eye(3)              # 3×3 单位矩阵
arange = torch.arange(0, 10, 2) # tensor([0, 2, 4, 6, 8])

# 创建与已有张量相同形状的张量
d = torch.zeros_like(c)         # 与 c 形状相同的全零张量
e = torch.ones_like(c)          # 与 c 形状相同的全一张量

tensor([1, 2, 3])
torch.Size([3])
torch.int64
torch.float32
torch.Size([2, 3])


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

#### 8.2.2 张量的基本运算

In [12]:
import torch

a = torch.tensor([1.0, 2.0, 3.0])
b = torch.tensor([4.0, 5.0, 6.0])

# 逐元素运算
print(a + b)      # tensor([5., 7., 9.])
print(a - b)      # tensor([-3., -3., -3.])
print(a * b)      # tensor([ 4., 10., 18.])  逐元素乘法
print(a / b)      # tensor([0.2500, 0.4000, 0.5000])
print(a ** 2)     # tensor([1., 4., 9.])

# 矩阵运算
A = torch.tensor([[1.0, 2.0], [3.0, 4.0]])
B = torch.tensor([[5.0, 6.0], [7.0, 8.0]])

print(A @ B)           # 矩阵乘法（等价于 torch.matmul(A, B)）
# tensor([[19., 22.],
#         [43., 50.]])

print(A.T)             # 矩阵转置
# tensor([[1., 3.],
#         [2., 4.]])

# 聚合运算
x = torch.tensor([1.0, 2.0, 3.0, 4.0, 5.0])
print(x.sum())         # tensor(15.)
print(x.mean())        # tensor(3.)
print(x.max())         # tensor(5.)
print(x.min())         # tensor(1.)
print(x.argmax())      # tensor(4)  最大值的索引

tensor([5., 7., 9.])
tensor([-3., -3., -3.])
tensor([ 4., 10., 18.])
tensor([0.2500, 0.4000, 0.5000])
tensor([1., 4., 9.])
tensor([[19., 22.],
        [43., 50.]])
tensor([[1., 3.],
        [2., 4.]])
tensor(15.)
tensor(3.)
tensor(5.)
tensor(1.)
tensor(4)


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

#### 8.2.3 张量的形状操作

In [13]:
import torch

x = torch.arange(12)
print(x)          # tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])
print(x.shape)    # torch.Size([12])

# reshape：改变形状
y = x.reshape(3, 4)
print(y)
# tensor([[ 0,  1,  2,  3],
#         [ 4,  5,  6,  7],
#         [ 8,  9, 10, 11]])

# view：改变形状（与 reshape 类似，但要求内存连续）
z = x.view(2, 6)
print(z.shape)    # torch.Size([2, 6])

# 使用 -1 自动推断维度
w = x.view(3, -1)  # -1 表示自动计算，等价于 view(3, 4)
print(w.shape)      # torch.Size([3, 4])

# 展平操作（CNN 中常用）
batch = torch.randn(16, 64, 10, 10)  # 批量大小=16, 通道=64, 高=10, 宽=10
flat = batch.view(batch.size(0), -1)  # 保持批量维度，展平其余维度
print(flat.shape)   # torch.Size([16, 6400])

# squeeze 和 unsqueeze
a = torch.randn(1, 3, 1)
print(a.shape)              # torch.Size([1, 3, 1])
print(a.squeeze().shape)    # torch.Size([3])  移除所有大小为1的维度
print(a.squeeze(0).shape)   # torch.Size([3, 1])  只移除第0维

b = torch.randn(3)
print(b.unsqueeze(0).shape) # torch.Size([1, 3])  在第0维添加维度
print(b.unsqueeze(1).shape) # torch.Size([3, 1])  在第1维添加维度

tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])
torch.Size([12])
tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])
torch.Size([2, 6])
torch.Size([3, 4])
torch.Size([16, 6400])
torch.Size([1, 3, 1])
torch.Size([3])
torch.Size([3, 1])
torch.Size([1, 3])
torch.Size([3, 1])


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

#### 8.2.4 数据类型转换

In [14]:
import torch

# 整数张量转浮点数
a = torch.tensor([1, 2, 3])
print(a.dtype)          # torch.int64
b = a.float()           # 转为 float32
print(b.dtype)          # torch.float32

# 也可以使用 .to() 方法
c = a.to(torch.float32)
d = a.to(torch.float64)

# 浮点数转整数
e = torch.tensor([1.7, 2.3, 3.9])
f = e.long()            # 转为 int64（截断小数部分）
print(f)                # tensor([1, 2, 3])

# 张量与 NumPy 数组互转
import numpy as np
np_array = np.array([1.0, 2.0, 3.0])
tensor_from_np = torch.from_numpy(np_array)    # NumPy → Tensor
np_from_tensor = tensor_from_np.numpy()        # Tensor → NumPy

# GPU 与 CPU 之间转移（如果有 GPU）
if torch.cuda.is_available():
    gpu_tensor = a.to('cuda')     # CPU → GPU
    cpu_tensor = gpu_tensor.to('cpu')  # GPU → CPU

torch.int64
torch.float32
tensor([1, 2, 3])


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

### 8.3 使用 PyTorch 构建神经网络

#### 8.3.1 nn.Module：网络的基类

PyTorch 中所有的神经网络都继承自 `torch.nn.Module`。构建网络需要：

1. 在 `__init__` 方法中定义网络的层
2. 在 `forward` 方法中定义前向传播的逻辑

In [15]:
import torch
import torch.nn as nn

class SimpleNet(nn.Module):
    def __init__(self):
        super(SimpleNet, self).__init__()
        self.fc1 = nn.Linear(784, 128)   # 全连接层：784 → 128
        self.relu = nn.ReLU()             # ReLU 激活函数
        self.fc2 = nn.Linear(128, 10)    # 全连接层：128 → 10
    
    def forward(self, x):
        x = self.fc1(x)    # 第一层线性变换
        x = self.relu(x)   # ReLU 激活
        x = self.fc2(x)    # 第二层线性变换
        return x

# 创建网络实例
model = SimpleNet()
print(model)

# 模拟输入（批量大小=4，输入维度=784）
x = torch.randn(4, 784)
output = model(x)
print(output.shape)   # torch.Size([4, 10])

SimpleNet(
  (fc1): Linear(in_features=784, out_features=128, bias=True)
  (relu): ReLU()
  (fc2): Linear(in_features=128, out_features=10, bias=True)
)
torch.Size([4, 10])


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

#### 8.3.2 nn.Sequential：快速构建网络

对于简单的顺序网络，可以使用 `nn.Sequential` 来简化代码：

In [16]:
import torch.nn as nn

model = nn.Sequential(
    nn.Linear(784, 128),
    nn.ReLU(),
    nn.Linear(128, 64),
    nn.ReLU(),
    nn.Linear(64, 10)
)

# 等价于上面的 SimpleNet，但不需要手动定义 forward 方法

Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

#### 8.3.3 常用层的 API

| 层 | PyTorch API | 关键参数 | 用途 |
|:---|:---|:---|:---|
| 全连接层 | `nn.Linear(in, out)` | in_features, out_features | 线性变换 |
| 卷积层 | `nn.Conv2d(in_ch, out_ch, kernel)` | in_channels, out_channels, kernel_size, stride, padding | 特征提取 |
| 最大池化 | `nn.MaxPool2d(kernel)` | kernel_size, stride | 下采样 |
| 平均池化 | `nn.AvgPool2d(kernel)` | kernel_size, stride | 下采样 |
| ReLU | `nn.ReLU()` | inplace | 激活函数 |
| Sigmoid | `nn.Sigmoid()` | — | 激活函数 |
| Dropout | `nn.Dropout(p)` | p（丢弃概率） | 正则化 |
| 批归一化 | `nn.BatchNorm2d(num_features)` | num_features | 加速训练 |

#### 8.3.4 构建 CNN（改编自 NOAI 2025 真题第 23 题）

In [17]:
import torch
import torch.nn as nn

class CaptchaCNN(nn.Module):
    def __init__(self):
        super(CaptchaCNN, self).__init__()
        
        # 第一个卷积块
        self.conv_block1 = nn.Sequential(
            nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1),  # 1×40×40 → 32×40×40
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2)                   # 32×40×40 → 32×20×20
        )
        
        # 第二个卷积块
        self.conv_block2 = nn.Sequential(
            nn.Conv2d(32, 64, kernel_size=5, stride=1, padding=2),  # 32×20×20 → 64×20×20
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2)                   # 64×20×20 → 64×10×10
        )
        
        # 全连接层
        self.classifier = nn.Sequential(
            nn.Linear(64 * 10 * 10, 128),  # 6400 → 128
            nn.ReLU(),
            nn.Linear(128, 26)              # 128 → 26（26个字母）
        )
    
    def forward(self, x):
        x = self.conv_block1(x)
        x = self.conv_block2(x)
        x = x.view(x.size(0), -1)   # 展平：保持批量维度，其余展平
        x = self.classifier(x)
        return x

# 测试
model = CaptchaCNN()
x = torch.randn(8, 1, 40, 40)  # 8张 40×40 灰度图像
output = model(x)
print(output.shape)  # torch.Size([8, 26])

torch.Size([8, 26])


Error: INVALID mime type: status. Must be in the format "type/subtype[;optionalparameter]"

### 8.4 PyTorch 训练流程

一个完整的 PyTorch 训练循环包含以下步骤：

In [20]:
# 本代码为框架，不可运行
import torch
import torch.nn as nn
import torch.optim as optim

# 1. 定义模型
model = SimpleNet()

# 2. 定义损失函数
criterion = nn.CrossEntropyLoss()   # 多分类交叉熵损失

# 3. 定义优化器
optimizer = optim.Adam(model.parameters(), lr=0.001)

# 4. 训练循环
for epoch in range(num_epochs):
    for batch_x, batch_y in dataloader:
        # 正向传播
        outputs = model(batch_x)
        loss = criterion(outputs, batch_y)
        
        # 反向传播
        optimizer.zero_grad()   # 清除之前的梯度
        loss.backward()         # 计算梯度
        optimizer.step()        # 更新参数
    
    print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")

训练流程中的关键步骤：

| 步骤 | 代码 | 作用 |
|:---|:---|:---|
| 正向传播 | `outputs = model(x)` | 计算预测输出 |
| 计算损失 | `loss = criterion(outputs, y)` | 衡量预测与真实的差距 |
| 清除梯度 | `optimizer.zero_grad()` | 防止梯度累加 |
| 反向传播 | `loss.backward()` | 计算所有参数的梯度 |
| 参数更新 | `optimizer.step()` | 根据梯度更新参数 |

### 章节练习

**【单选题 1】** 在 PyTorch 中，以下哪个方法用于将多维张量展平为一维？

A. `x.flatten()`
B. `x.squeeze()`
C. `x.unsqueeze(0)`
D. `x.transpose(0, 1)`

**【单选题 2】** 在 PyTorch 的训练循环中，`optimizer.zero_grad()` 的作用是什么？

A. 将模型参数初始化为零
B. 清除之前计算的梯度，防止梯度累加
C. 将学习率设置为零
D. 将损失值重置为零

**【代码填空】** 以下代码使用 PyTorch 构建一个简单的 CNN。请补全标记处的代码。（改编自 NOAI 2025 真题第 23 题）

```python
import torch
import torch.nn as nn

class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()
        # 卷积块：输入1通道，输出32通道，卷积核3×3
        self.conv1 = ______(1)______
        self.relu = nn.ReLU()
        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
        # 全连接层
        self.fc1 = nn.Linear(32 * 14 * 14, 10)
    
    def forward(self, x):
        x = self.conv1(x)       # 卷积
        x = self.relu(x)        # 激活
        x = self.pool(x)        # 池化
        x = ______(2)______     # 展平
        x = self.fc1(x)         # 全连接
        return x
```

## 第九章 综合练习与真题演练

本章汇集了模拟 NOAI 真题风格的综合练习题，涵盖本教程的所有核心知识点。

### 9.1 选择题

**【第 1 题】** 以下关于感知机的描述，哪一项是错误的？

A. 感知机可以实现逻辑与（AND）运算
B. 感知机可以实现逻辑或（OR）运算
C. 单层感知机可以实现异或（XOR）运算
D. 感知机的输出由阶跃函数决定

**【第 2 题】** 在训练神经网络时，以下哪个步骤不属于反向传播阶段？

A. 计算损失函数对输出层权重的梯度
B. 将误差信号从输出层传播到隐藏层
C. 根据当前参数计算网络的预测输出
D. 利用链式法则计算各层参数的偏导数

**【第 3 题】** 以下哪种优化器为每个参数自适应调整学习率？

A. 批量梯度下降（Batch GD）
B. 随机梯度下降（SGD）
C. Adam
D. 小批量梯度下降（Mini-batch GD）

**【第 4 题】** 一个 $8 \times 8$ 的输入图像，使用 $3 \times 3$ 的卷积核，步幅为 2，填充为 1，输出特征图的尺寸是？

A. $3 \times 3$
B. $4 \times 4$
C. $5 \times 5$
D. $6 \times 6$

**【第 5 题】** 以下关于池化层的描述，哪一项是正确的？

A. 池化层包含可学习的参数
B. 最大池化取窗口内的平均值
C. 池化层可以增大特征图的空间尺寸
D. 池化层没有可学习的参数

**【第 6 题】** 大语言模型（LLM）的核心训练目标是什么？

A. 对输入文本进行语法纠错
B. 给定前文，预测下一个最可能出现的词
C. 将文本翻译成多种语言
D. 从文本中提取关键词

**【第 7 题】** 在 PyTorch 中，`torch.randn(3, 4)` 创建的张量形状是？

A. 一个包含 3 个元素的一维张量
B. 一个 3×4 的二维张量
C. 一个 4×3 的二维张量
D. 一个包含 12 个元素的一维张量

**【第 8 题】** 以下哪种数据增强方法不适用于图像数据？

A. 水平翻转
B. 随机裁剪
C. 前向填充
D. 添加高斯噪声

### 9.2 简答题

**【第 9 题】** 请简要描述卷积神经网络（CNN）中卷积层、池化层和全连接层各自的作用。

**【第 10 题】** 请解释 Adam 优化器与普通随机梯度下降（SGD）的主要区别，并说明 Adam 的核心优势。

**【第 11 题】** 请简要描述 Transformer 架构中自注意力机制的核心思想，以及它相比 RNN 的优势。

### 9.3 代码填空题

**【第 12 题】** 以下代码使用 PyTorch 定义了一个简单的全连接神经网络并完成一次前向传播。请补全标记处的代码。

```python
import torch
import torch.nn as nn

class MyNet(nn.Module):
    def __init__(self):
        super(MyNet, self).__init__()
        self.fc1 = nn.Linear(10, 32)
        self.relu = ______(1)______
        self.fc2 = nn.Linear(32, 5)
    
    def forward(self, x):
        x = self.fc1(x)
        x = ______(2)______
        x = self.fc2(x)
        return x

model = MyNet()
x = torch.randn(______(3)______)   # 创建一个批量大小为4、特征维度为10的输入
output = model(x)
print(output.shape)  # 期望输出: torch.Size([4, 5])
```

**【第 13 题】** 以下代码实现了一个完整的 PyTorch 训练步骤。请补全标记处的代码。

```python
import torch
import torch.nn as nn
import torch.optim as optim

model = nn.Linear(10, 2)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# 模拟数据
x = torch.randn(32, 10)
y = torch.randint(0, 2, (32,))

# 训练一步
output = model(x)
loss = criterion(output, y)

______(1)______    # 清除之前的梯度
______(2)______    # 反向传播计算梯度
______(3)______    # 更新参数
```

**【第 14 题】** 以下代码实现了文本的 tokenization 和词汇表构建。请补全标记处的代码。

```python
def build_vocab(texts):
    vocab = {'<PAD>': 0, '<UNK>': 1}
    idx = 2
    for text in texts:
        for word in text.lower().split():
            if word not in vocab:
                ______(1)______   # 将新词添加到词汇表
                idx += 1
    return vocab

def encode(text, vocab, max_len=10):
    tokens = text.lower().split()
    indices = [______(2)______ for token in tokens]  # 将每个词转为索引，未知词用 <UNK> 的索引
    # 填充到固定长度
    while len(indices) < max_len:
        indices.append(______(3)______)  # 用 PAD 的索引填充
    return indices[:max_len]
```

**【第 15 题】** 以下代码对图像进行卷积运算（不使用深度学习框架）。请补全标记处的代码。（改编自 NOAI 2025 真题第 22 题风格）

```python
import numpy as np

def conv2d(image, kernel):
    """手动实现二维卷积"""
    img_h, img_w = image.shape
    k_h, k_w = kernel.shape
    
    # 计算输出尺寸
    out_h = ______(1)______
    out_w = ______(2)______
    
    output = np.zeros((out_h, out_w))
    
    for i in range(out_h):
        for j in range(out_w):
            # 提取对应区域并计算内积
            region = image[i:i+k_h, j:j+k_w]
            output[i, j] = ______(3)______
    
    return output

# 测试
image = np.array([
    [1, 1, 1],
    [1, 2, 1],
    [1, 1, 1]
], dtype=float)

kernel = np.array([
    [1,  1],
    [1, -1]
], dtype=float)

result = conv2d(image, kernel)
print(result)
# 期望输出: [[1. 3.]
#            [3. 3.]]
```

---

## 附录A 参考答案与解析

### 第一章 参考答案

**【单选题 1】** 答案：**D**

解析：单层感知机只能解决线性可分问题。AND、OR、NOT 都是线性可分的，可以用一条直线分割两类数据。而 XOR 是线性不可分的，需要多层网络才能解决。

**【单选题 2】** 答案：**B**

解析：权重矩阵的维度由"目标层神经元数 × 源层神经元数"决定。从输入层（4 个神经元）到隐藏层（6 个神经元），权重矩阵的维度为 6 × 4。这是因为隐藏层的每个神经元都需要与输入层的每个神经元相连，所以需要 6 行 4 列的矩阵。

**【简答题】** 参考答案：

单层感知机本质上是在特征空间中画一条直线（或超平面）来分割两类数据。对于 XOR 问题，正类 (0,1) 和 (1,0) 与负类 (0,0) 和 (1,1) 在二维平面上无法用一条直线分开，即数据是线性不可分的，因此单层感知机无法解决。

多层神经网络通过引入隐藏层和非线性激活函数来克服这一局限。隐藏层对原始输入进行非线性变换，将数据映射到一个新的特征空间。在这个新空间中，原本线性不可分的数据变得线性可分，输出层就可以用一条直线将两类数据分开。

### 第二章 参考答案

**【单选题 1】** 答案：**D**

解析：反向传播算法的核心作用是利用链式法则计算损失函数对网络中所有权重和偏置项的梯度，然后根据这些梯度来调整（更新）网络的参数。选项 A 描述的是正向传播，选项 B 是网络设计阶段的任务，选项 C 是模型评估的任务。

**【单选题 2】** 答案：**B**

解析：反向传播的数学基础是微积分中的链式法则（Chain Rule）。由于损失函数通过多层嵌套函数与各层参数相关联，链式法则允许我们将复杂的偏导数分解为一系列简单偏导数的乘积，从而逐层计算梯度。

**【简答题】** 参考答案：

正向传播是数据从输入层经过隐藏层到达输出层的过程。在每一层中，神经元接收上一层的输出，进行加权求和、加偏置、通过激活函数处理，然后将结果传递给下一层，最终在输出层产生预测结果。

反向传播是从输出层开始，利用链式法则逐层向后计算损失函数对每个参数（权重和偏置）的偏导数（梯度）的过程。它首先计算输出层的误差信号，然后将这个误差信号反向传播到每一个隐藏层，计算出各层参数的梯度。

两者的关系是：正向传播产生预测结果并计算损失值，反向传播根据损失值计算梯度，然后使用梯度下降法更新参数。训练过程就是不断交替执行正向传播和反向传播，逐步优化网络参数，使损失值不断减小。

### 第三章 参考答案

**【单选题 1】** 答案：**C**

解析：梯度下降法的主要目的是通过沿梯度反方向迭代更新参数，找到损失函数的最小值。这是机器学习中最基本的优化方法。

**【单选题 2】** 答案：**C**

解析：Adam 优化器的核心特点是结合了动量法（一阶矩估计）和自适应学习率（二阶矩估计）。它为每个参数维护独立的学习率，并利用历史梯度信息来加速收敛。

**【代码填空】** 答案：

- (1) `diff * x[i]`（损失对 w 的偏导数：误差乘以对应的输入值）
- (2) `diff`（损失对 b 的偏导数：就是误差本身）
- (3) `w - alpha * dw`（沿梯度反方向更新 w）
- (4) `b - alpha * db`（沿梯度反方向更新 b）

### 第四章 参考答案

**【单选题 1】** 答案：**C**

解析：CNN 前向传播的典型层顺序是：卷积层（提取特征）→ 池化层（降低维度）→ 全连接层（分类/回归）→ 输出层。这个顺序在 NOAI 2025 真题中直接考查。

**【单选题 2】** 答案：**B**

解析：矩阵内积 $A \odot B = 1 \times 1 + 1 \times 2 + 1 \times 3 + (-1) \times 4 = 1 + 2 + 3 - 4 = 2$。注意这里的内积是逐元素相乘再求和，不是矩阵乘法。

**【单选题 3】** 答案：**C**

解析：根据输出尺寸公式：$\lfloor(6 - 3 + 2 \times 0) / 1\rfloor + 1 = \lfloor 3/1 \rfloor + 1 = 4$。输出为 $4 \times 4$。

**【代码填空】** 答案：

- (1) `image[i + di][j + dj]`（访问邻居像素值）
- (2) `tmp / count`（计算邻居像素的平均值）

### 第五章 参考答案

**【单选题 1】** 答案：**C**

解析：Softmax 将多个输出值转化为概率分布，是决策函数；ReLU 引入非线性，是激活函数；交叉熵计算预测与真实标签的误差，是损失函数。

**【单选题 2】** 答案：**B**

解析：激活函数的作用就是将线性求和的结果进行非线性映射，使得神经网络能够学习非线性关系。

**【代码填空】** 答案：

- (1) `np.exp(z)`（计算每个元素的指数值）
- (2) `np.sum(exp_z)`（计算所有指数值的总和）

### 第六章 参考答案

**【单选题 1】** 答案：**B**

解析：LLM 的核心训练目标是自回归语言建模，即给定前文预测下一个最可能出现的词。通过在海量文本上训练这个简单任务，模型逐渐学会了语言规律和世界知识。

**【单选题 2】** 答案：**C**

解析：Transformer 架构的核心创新是自注意力机制（Self-Attention），它允许模型在处理序列中的每个位置时同时关注所有其他位置的信息，且可以并行计算。

**【简答题】** 参考答案：

大语言模型的核心思想是自回归语言建模：给定一段文本的前文，预测下一个最可能出现的词（token）。模型在海量文本数据上反复训练这个任务，逐渐学会语言的语法、语义、逻辑推理和世界知识。生成文本时，模型逐词预测并将结果添加到序列末尾，循环往复生成完整文本。

涌现能力是指在小规模模型上完全不具备、但当模型规模（参数量、数据量、计算量）增大到一定程度时突然出现的能力。例如，小模型无法进行复杂的逻辑推理或代码生成，但当模型参数量达到数百亿级别时，这些能力会突然涌现。涌现能力的出现表明，大语言模型的能力并非简单的线性增长，而是存在质变的临界点。

### 第七章 参考答案（Python 面向对象编程）

**【单选题 1】** 答案：**B**

解析：`__init__` 是 Python 类的构造方法，在创建对象时会自动被调用。`__call__` 是在对象被像函数一样调用时触发，`__repr__` 是在打印对象时触发，`forward` 是 PyTorch 中的自定义方法，不是 Python 内置的特殊方法。

**【单选题 2】** 答案：**A**

解析：在 PyTorch 中，自定义模型类必须继承 `nn.Module` 并重写 `forward` 方法来定义前向传播逻辑。同时必须在 `__init__` 中调用 `super().__init__()` 来初始化父类。网络层应在 `__init__` 中定义，而不是 `forward` 中。

**【代码填空】** 答案：

- (1) `nn.Module`（继承 PyTorch 的基础模型类）
- (2) `super().__init__()`（调用父类的构造方法）
- (3) `forward`（定义前向传播方法）

### 第八章 参考答案（PyTorch 框架基础）

**【单选题 1】** 答案：**A**

解析：`x.flatten()` 将多维张量展平为一维。`squeeze()` 是移除大小为 1 的维度，`unsqueeze()` 是添加维度，`transpose()` 是转置。

**【单选题 2】** 答案：**B**

解析：`optimizer.zero_grad()` 的作用是清除之前计算的梯度。PyTorch 默认会累加梯度，如果不清除，新计算的梯度会与之前的梯度叠加，导致参数更新不正确。

**【代码填空】** 答案：

- (1) `nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1)`（定义卷积层：1 个输入通道，32 个输出通道，3×3 卷积核，padding=1 保持尺寸）
- (2) `x.view(x.size(0), -1)`（展平操作：保持批量维度，其余维度展平为一维）


### 第十章 综合练习参考答案

**【第 1 题】** 答案：**C**。单层感知机无法实现 XOR，需要多层网络。

**【第 2 题】** 答案：**C**。根据当前参数计算预测输出属于正向传播阶段，不属于反向传播。

**【第 3 题】** 答案：**C**。Adam 通过二阶矩估计为每个参数自适应调整学习率。

**【第 4 题】** 答案：**B**。$\lfloor(8 - 3 + 2 \times 1) / 2\rfloor + 1 = \lfloor 7/2 \rfloor + 1 = 3 + 1 = 4$。

**【第 5 题】** 答案：**D**。池化层是固定的下采样操作，没有可学习的参数。

**【第 6 题】** 答案：**B**。LLM 的核心目标是自回归语言建模——预测下一个词。

**【第 7 题】** 答案：**B**。`torch.randn(3, 4)` 创建一个 3 行 4 列的二维张量。

**【第 8 题】** 答案：**C**。前向填充是处理时间序列缺失值的方法，不是图像数据增强方法。

**【第 9 题】** 参考答案：

卷积层的作用是通过卷积核在输入上滑动，提取局部特征（如边缘、纹理等）。卷积层利用局部连接和权值共享大大减少了参数数量。

池化层的作用是降低特征图的空间尺寸，减少计算量和参数数量，同时增强特征的平移不变性。常见的池化方式有最大池化和平均池化，池化层没有可学习的参数。

全连接层的作用是将卷积和池化层提取的特征映射到最终的输出。在进入全连接层之前，需要将多维特征图展平为一维向量。全连接层通常位于网络末端，负责最终的分类或回归决策。

**【第 10 题】** 参考答案：

普通 SGD 对所有参数使用相同的固定学习率，仅根据当前梯度更新参数。Adam 的主要区别在于：（1）引入一阶矩估计（动量），利用历史梯度的指数移动平均来提供更新方向的惯性，有助于加速收敛和跳出局部最优；（2）引入二阶矩估计（自适应学习率），跟踪梯度平方的指数移动平均，为每个参数自动调整学习率——梯度大的参数学习率自动减小，梯度小的参数学习率自动增大。Adam 的核心优势是对超参数不敏感（默认参数通常就能工作良好），收敛速度快，适用于大多数深度学习任务。

**【第 11 题】** 参考答案：

自注意力机制的核心思想是：在处理序列中的每个位置时，通过计算该位置与序列中所有其他位置的相关性（注意力分数），来决定应该关注哪些位置的信息。具体实现通过查询（Q）、键（K）、值（V）三个矩阵：用 Q 和 K 的点积计算注意力分数，经过 Softmax 归一化后，对 V 进行加权求和，得到融合了全局信息的输出。

相比 RNN，自注意力机制的主要优势是：（1）可以并行处理整个序列，而 RNN 必须逐步处理，训练效率大大提高；（2）任意两个位置之间的信息传递路径长度为 O(1)，而 RNN 为 O(n)，更容易捕捉长距离依赖关系。

**【第 12 题】** 答案：

- (1) `nn.ReLU()`
- (2) `self.relu(x)`
- (3) `4, 10`

**【第 13 题】** 答案：

- (1) `optimizer.zero_grad()`
- (2) `loss.backward()`
- (3) `optimizer.step()`

**【第 14 题】** 答案：

- (1) `vocab[word] = idx`
- (2) `vocab.get(token, vocab['<UNK>'])`
- (3) `vocab['<PAD>']`（即 `0`）

**【第 15 题】** 答案：

- (1) `img_h - k_h + 1`
- (2) `img_w - k_w + 1`
- (3) `np.sum(region * kernel)`

## 附录B 参考资源

### 一、官方文档与框架

1. **PyTorch 官方文档**. PyTorch Documentation. https://pytorch.org/docs/stable/
2. **PyTorch 官方教程**. PyTorch Tutorials. https://pytorch.org/tutorials/
3. **NumPy 官方文档**. NumPy Documentation. https://numpy.org/doc/stable/
4. **Scikit-Learn 官方文档**. Scikit-Learn Documentation. https://scikit-learn.org/stable/
5. **Python 官方文档**. Python Documentation. https://docs.python.org/3/

### 二、经典教材与书籍

6. Goodfellow, I., Bengio, Y., & Courville, A. (2016). *Deep Learning*. MIT Press. https://www.deeplearningbook.org/
7. Bishop, C. M. (2006). *Pattern Recognition and Machine Learning*. Springer.
8. 周志华. (2016). *机器学习*. 清华大学出版社.
9. 李航. (2012). *统计学习方法*. 清华大学出版社.
10. Raschka, S., & Mirjalili, V. (2019). *Python Machine Learning* (3rd ed.). Packt Publishing.

### 三、在线课程与学习资源

11. **Stanford CS231n**: Convolutional Neural Networks for Visual Recognition. https://cs231n.stanford.edu/
12. **Stanford CS229**: Machine Learning. https://cs229.stanford.edu/
13. **3Blue1Brown**: Neural Networks (YouTube 系列视频). https://www.3blue1brown.com/topics/neural-networks
14. **Andrew Ng 机器学习课程**. Coursera Machine Learning Specialization. https://www.coursera.org/specializations/machine-learning-introduction
15. **动手学深度学习** (Dive into Deep Learning). 李沐 等. https://d2l.ai/

### 四、研究论文

16. Rosenblatt, F. (1958). The Perceptron: A Probabilistic Model for Information Storage and Organization in the Brain. *Psychological Review*, 65(6), 386–408.
17. Rumelhart, D. E., Hinton, G. E., & Williams, R. J. (1986). Learning representations by back-propagating errors. *Nature*, 323(6088), 533–536.
18. LeCun, Y., Bottou, L., Bengio, Y., & Haffner, P. (1998). Gradient-based learning applied to document recognition. *Proceedings of the IEEE*, 86(11), 2278–2324.
19. Kingma, D. P., & Ba, J. (2015). Adam: A Method for Stochastic Optimization. *Proceedings of ICLR 2015*. arXiv:1412.6980.
20. Loshchilov, I., & Hutter, F. (2019). Decoupled Weight Decay Regularization. *Proceedings of ICLR 2019*. arXiv:1711.05101.
21. Vaswani, A., Shazeer, N., Parmar, N., et al. (2017). Attention Is All You Need. *Advances in Neural Information Processing Systems (NeurIPS)*, 30. arXiv:1706.03762.
22. Dosovitskiy, A., Beyer, L., Kolesnikov, A., et al. (2021). An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale. *Proceedings of ICLR 2021*. arXiv:2010.11929.
23. He, K., Zhang, X., Ren, S., & Sun, J. (2016). Deep Residual Learning for Image Recognition. *Proceedings of CVPR 2016*. arXiv:1512.03385.
24. Krizhevsky, A., Sutskever, I., & Hinton, G. E. (2012). ImageNet Classification with Deep Convolutional Neural Networks. *Advances in Neural Information Processing Systems (NeurIPS)*, 25.

### 五、NOAI 官方资源

25. **NOAI 2024 水平测试试题与答案**. NOAI 组委会.
26. **NOAI 2025 水平测试试题与答案**. NOAI 组委会.
27. **NOAI 2026 学术大纲**. NOAI 组委会. 

### 六、实践工具与平台

这些平台和工具仅供参考，NOAI竞赛目前采用的实验环境是[玻尔®️ / Bohrium®️平台](https://www.bohrium.com/)

28. **Google Colab**: 免费的云端 Jupyter Notebook 环境，支持 GPU. https://colab.research.google.com/
29. **Kaggle**: 数据科学竞赛平台，提供免费计算资源和数据集. https://www.kaggle.com/
30. **Hugging Face**: 预训练模型和数据集中心. https://huggingface.co/
31. **Papers with Code**: 论文与代码实现的对应平台. https://paperswithcode.com/